<!DOCTYPE HTML>
<html lang="en">
<head>
<!-- Generated by javadoc (18) on Fri Feb 24 15:31:35 EST 2023 -->
<title>PBILSpecies (ecj 27 API)</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="dc.created" content="2023-02-24">
<meta name="description" content="declaration: package: ec.eda.pbil, class: PBILSpecies">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var evenRowColor = "even-row-color";
var oddRowColor = "odd-row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../../../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar-top"><button id="navbar-toggle-button" aria-controls="navbar-top" aria-expanded="false" aria-label="Toggle navigation links"><span class="nav-bar-toggle-icon"></span><span class="nav-bar-toggle-icon"></span><span class="nav-bar-toggle-icon"></span></button>
<div class="skip-nav"><a href="#skip-navbar-top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar-top-firstrow" class="nav-list" title="Navigation">
<li><a href="../../../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="class-use/PBILSpecies.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-all.html">Index</a></li>
<li><a href="../../../help-doc.html#class">Help</a></li>
</ul>
<ul class="sub-nav-list-small">
<li>
<p>Summary:</p>
<ul>
<li>Nested</li>
<li><a href="#field-summary">Field</a></li>
<li><a href="#constructor-summary">Constr</a></li>
<li><a href="#method-summary">Method</a></li>
</ul>
</li>
<li>
<p>Detail:</p>
<ul>
<li><a href="#field-detail">Field</a></li>
<li><a href="#constructor-detail">Constr</a></li>
<li><a href="#method-detail">Method</a></li>
</ul>
</li>
</ul>
</div>
<div class="sub-nav">
<div id="navbar-sub-list">
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field-summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor-summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method-summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field-detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor-detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method-detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search-input">SEARCH:</label>
<input type="text" id="search-input" disabled placeholder="Search">
<input type="reset" id="reset-button" disabled value="reset">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip-navbar-top"></span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">ec.eda.pbil</a></div>
<h1 title="Class PBILSpecies" class="title">Class PBILSpecies</h1>
</div>
<div class="inheritance" title="Inheritance Tree"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">java.lang.Object</a>
<div class="inheritance"><a href="../../Species.html" title="class in ec">ec.Species</a>
<div class="inheritance"><a href="../../vector/VectorSpecies.html" title="class in ec.vector">ec.vector.VectorSpecies</a>
<div class="inheritance"><a href="../../vector/IntegerVectorSpecies.html" title="class in ec.vector">ec.vector.IntegerVectorSpecies</a>
<div class="inheritance">ec.eda.pbil.PBILSpecies</div>
</div>
</div>
</div>
</div>
<section class="class-description" id="class-description">
<dl class="notes">
<dt>All Implemented Interfaces:</dt>
<dd><code><a href="../../Prototype.html" title="interface in ec">Prototype</a></code>, <code><a href="../../Setup.html" title="interface in ec">Setup</a></code>, <code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/io/Serializable.html" title="class or interface in java.io" class="external-link">Serializable</a></code>, <code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Cloneable.html" title="class or interface in java.lang" class="external-link">Cloneable</a></code></dd>
</dl>
<hr>
<div class="type-signature"><span class="modifiers">public class </span><span class="element-name type-name-label">PBILSpecies</span>
<span class="extends-implements">extends <a href="../../vector/IntegerVectorSpecies.html" title="class in ec.vector">IntegerVectorSpecies</a></span></div>
<div class="block">PBILSpecies is an IntegerVectorSpecies which implements a faithful version of the
 PBIL algorithm.  The class has two basic methods.  The newIndividual(...)
 method generates a new random individual underneath the current PBIL marginal
 distribution.  The updateDistribution(...) method revises the marginal 
 distribution to reflect the fitness results of the population.
 
 <p>PBILSpecies must be used in combination with PBILBreeder, which will
 call it at appropriate times to revise the distribution and to generate a new
 subpopulation of individuals.  Since the initial population is built based on
 the marginal distributions, SimpleInitializer is used to generate the initial 
 population. 

 <p>PBILSpecies <b>needs the population size and also truncation size</b>. The
 truncation size which is the 'b' parameter, denotes how many fittest 
 individuals to pick out from the generated population of individuals. 
 It also needs the genome size and the minimum and maximum values of the genes. 
 The size of the range of min and max gene values should be specified for each 
 gene in the parameters file. 

 <p>PBILSpecies also uses the learning rate 'alpha' based on which it decides 
 the amount of old distribution to be retained and how much of the new 
 distribution to be added. 'alpha' and 'b' values are printed out when running
 so the user may see what values it used for that given run.  


 <p><b>Parameters</b><br>
 <table>
 <tr><td valign=top><i>base</i>.<tt>alpha</tt><br>
 <font size=-1>Floating-point value ranging from 0 to 1.0 </font></td>
 <td valign=top>(the learning rate parameter)<br> 
 If not provided, this defaults to 0.05.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>popsize</tt><br>
 <font size=-1>Integer > 0 </font></td>
 <td valign=top>(pop.subpop.0.size population size)<br>
 This parameter must be provided.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>b</tt><br>
 <font size=-1>Integer ranging from 1 to popsize </font></td>
 <td valign=top>(b truncated population size)<br>
 This parameter must be provided.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>n</tt><br>
 <font size=-1>Integer > 1</td>
 <td valign=top>(n genome size)<br>
 This parameter must be provided. 
 </td></tr>

 </table>


 <p><b>Default Base</b><br>
 eda.pbil.species</div>
<dl class="notes">
<dt>Version:</dt>
<dd>1.0</dd>
<dt>Author:</dt>
<dd>Sunil Kumar Rajendran</dd>
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../serialized-form.html#ec.eda.pbil.PBILSpecies">Serialized Form</a></li>
</ul>
</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field-summary">
<h2>Field Summary</h2>
<div class="caption"><span>Fields</span></div>
<div class="summary-table three-column-summary">
<div class="table-header col-first">Modifier and Type</div>
<div class="table-header col-second">Field</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color"><code>double</code></div>
<div class="col-second even-row-color"><code><a href="#alpha" class="member-name-link">alpha</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>int</code></div>
<div class="col-second odd-row-color"><code><a href="#b" class="member-name-link">b</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>private <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/util/List.html" title="class or interface in java.util" class="external-link">List</a>&lt;double[]&gt;</code></div>
<div class="col-second even-row-color"><code><a href="#distributions" class="member-name-link">distributions</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_ALPHA" class="member-name-link">P_ALPHA</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_B" class="member-name-link">P_B</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_PBIL_SPECIES" class="member-name-link">P_PBIL_SPECIES</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
</div>
<div class="inherited-list">
<h3 id="fields-inherited-from-class-ec.vector.IntegerVectorSpecies">Fields inherited from class&nbsp;ec.vector.<a href="../../vector/IntegerVectorSpecies.html" title="class in ec.vector">IntegerVectorSpecies</a></h3>
<code><a href="../../vector/IntegerVectorSpecies.html#C_RANDOM_WALK_MUTATION">C_RANDOM_WALK_MUTATION</a>, <a href="../../vector/IntegerVectorSpecies.html#C_RESET_MUTATION">C_RESET_MUTATION</a>, <a href="../../vector/IntegerVectorSpecies.html#maxGene">maxGene</a>, <a href="../../vector/IntegerVectorSpecies.html#minGene">minGene</a>, <a href="../../vector/IntegerVectorSpecies.html#mutationIsBounded">mutationIsBounded</a>, <a href="../../vector/IntegerVectorSpecies.html#mutationType">mutationType</a>, <a href="../../vector/IntegerVectorSpecies.html#P_MAXGENE">P_MAXGENE</a>, <a href="../../vector/IntegerVectorSpecies.html#P_MINGENE">P_MINGENE</a>, <a href="../../vector/IntegerVectorSpecies.html#P_MUTATION_BOUNDED">P_MUTATION_BOUNDED</a>, <a href="../../vector/IntegerVectorSpecies.html#P_MUTATIONTYPE">P_MUTATIONTYPE</a>, <a href="../../vector/IntegerVectorSpecies.html#P_NUM_SEGMENTS">P_NUM_SEGMENTS</a>, <a href="../../vector/IntegerVectorSpecies.html#P_RANDOM_WALK_PROBABILITY">P_RANDOM_WALK_PROBABILITY</a>, <a href="../../vector/IntegerVectorSpecies.html#P_SEGMENT">P_SEGMENT</a>, <a href="../../vector/IntegerVectorSpecies.html#P_SEGMENT_END">P_SEGMENT_END</a>, <a href="../../vector/IntegerVectorSpecies.html#P_SEGMENT_START">P_SEGMENT_START</a>, <a href="../../vector/IntegerVectorSpecies.html#P_SEGMENT_TYPE">P_SEGMENT_TYPE</a>, <a href="../../vector/IntegerVectorSpecies.html#randomWalkProbability">randomWalkProbability</a>, <a href="../../vector/IntegerVectorSpecies.html#V_RANDOM_WALK_MUTATION">V_RANDOM_WALK_MUTATION</a>, <a href="../../vector/IntegerVectorSpecies.html#V_RESET_MUTATION">V_RESET_MUTATION</a></code></div>
<div class="inherited-list">
<h3 id="fields-inherited-from-class-ec.vector.VectorSpecies">Fields inherited from class&nbsp;ec.vector.<a href="../../vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></h3>
<code><a href="../../vector/VectorSpecies.html#C_ANY_POINT">C_ANY_POINT</a>, <a href="../../vector/VectorSpecies.html#C_GEOMETRIC">C_GEOMETRIC</a>, <a href="../../vector/VectorSpecies.html#C_INTERMED_RECOMB">C_INTERMED_RECOMB</a>, <a href="../../vector/VectorSpecies.html#C_LINE_RECOMB">C_LINE_RECOMB</a>, <a href="../../vector/VectorSpecies.html#C_NONE">C_NONE</a>, <a href="../../vector/VectorSpecies.html#C_ONE_POINT">C_ONE_POINT</a>, <a href="../../vector/VectorSpecies.html#C_ONE_POINT_NO_NOP">C_ONE_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#C_SIMULATED_BINARY">C_SIMULATED_BINARY</a>, <a href="../../vector/VectorSpecies.html#C_TWO_POINT">C_TWO_POINT</a>, <a href="../../vector/VectorSpecies.html#C_TWO_POINT_NO_NOP">C_TWO_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#C_UNIFORM">C_UNIFORM</a>, <a href="../../vector/VectorSpecies.html#chunksize">chunksize</a>, <a href="../../vector/VectorSpecies.html#crossoverDistributionIndex">crossoverDistributionIndex</a>, <a href="../../vector/VectorSpecies.html#crossoverProbability">crossoverProbability</a>, <a href="../../vector/VectorSpecies.html#crossoverType">crossoverType</a>, <a href="../../vector/VectorSpecies.html#duplicateRetries">duplicateRetries</a>, <a href="../../vector/VectorSpecies.html#dynamicInitialSize">dynamicInitialSize</a>, <a href="../../vector/VectorSpecies.html#genomeIncreaseProbability">genomeIncreaseProbability</a>, <a href="../../vector/VectorSpecies.html#genomeResizeAlgorithm">genomeResizeAlgorithm</a>, <a href="../../vector/VectorSpecies.html#genomeSize">genomeSize</a>, <a href="../../vector/VectorSpecies.html#lineDistance">lineDistance</a>, <a href="../../vector/VectorSpecies.html#maxInitialSize">maxInitialSize</a>, <a href="../../vector/VectorSpecies.html#minInitialSize">minInitialSize</a>, <a href="../../vector/VectorSpecies.html#mutationProbability">mutationProbability</a>, <a href="../../vector/VectorSpecies.html#P_CHUNKSIZE">P_CHUNKSIZE</a>, <a href="../../vector/VectorSpecies.html#P_CROSSOVER_DISTRIBUTION_INDEX">P_CROSSOVER_DISTRIBUTION_INDEX</a>, <a href="../../vector/VectorSpecies.html#P_CROSSOVERPROB">P_CROSSOVERPROB</a>, <a href="../../vector/VectorSpecies.html#P_CROSSOVERTYPE">P_CROSSOVERTYPE</a>, <a href="../../vector/VectorSpecies.html#P_DUPLICATE_RETRIES">P_DUPLICATE_RETRIES</a>, <a href="../../vector/VectorSpecies.html#P_GENOMESIZE">P_GENOMESIZE</a>, <a href="../../vector/VectorSpecies.html#P_GEOMETRIC_PROBABILITY">P_GEOMETRIC_PROBABILITY</a>, <a href="../../vector/VectorSpecies.html#P_LINEDISTANCE">P_LINEDISTANCE</a>, <a href="../../vector/VectorSpecies.html#P_MUTATIONPROB">P_MUTATIONPROB</a>, <a href="../../vector/VectorSpecies.html#P_UNIFORM_MAX">P_UNIFORM_MAX</a>, <a href="../../vector/VectorSpecies.html#P_UNIFORM_MIN">P_UNIFORM_MIN</a>, <a href="../../vector/VectorSpecies.html#P_VECTORSPECIES">P_VECTORSPECIES</a>, <a href="../../vector/VectorSpecies.html#V_ANY_POINT">V_ANY_POINT</a>, <a href="../../vector/VectorSpecies.html#V_GEOMETRIC">V_GEOMETRIC</a>, <a href="../../vector/VectorSpecies.html#V_INTERMED_RECOMB">V_INTERMED_RECOMB</a>, <a href="../../vector/VectorSpecies.html#V_LINE_RECOMB">V_LINE_RECOMB</a>, <a href="../../vector/VectorSpecies.html#V_ONE_POINT">V_ONE_POINT</a>, <a href="../../vector/VectorSpecies.html#V_ONE_POINT_NO_NOP">V_ONE_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#V_SIMULATED_BINARY">V_SIMULATED_BINARY</a>, <a href="../../vector/VectorSpecies.html#V_TWO_POINT">V_TWO_POINT</a>, <a href="../../vector/VectorSpecies.html#V_TWO_POINT_NO_NOP">V_TWO_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#V_UNIFORM">V_UNIFORM</a></code></div>
<div class="inherited-list">
<h3 id="fields-inherited-from-class-ec.Species">Fields inherited from class&nbsp;ec.<a href="../../Species.html" title="class in ec">Species</a></h3>
<code><a href="../../Species.html#f_prototype">f_prototype</a>, <a href="../../Species.html#i_prototype">i_prototype</a>, <a href="../../Species.html#P_FITNESS">P_FITNESS</a>, <a href="../../Species.html#P_INDIVIDUAL">P_INDIVIDUAL</a>, <a href="../../Species.html#P_PIPE">P_PIPE</a>, <a href="../../Species.html#pipe_prototype">pipe_prototype</a></code></div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor-summary">
<h2>Constructor Summary</h2>
<div class="caption"><span>Constructors</span></div>
<div class="summary-table two-column-summary">
<div class="table-header col-first">Constructor</div>
<div class="table-header col-last">Description</div>
<div class="col-constructor-name even-row-color"><code><a href="#%3Cinit%3E()" class="member-name-link">PBILSpecies</a>()</code></div>
<div class="col-last even-row-color">&nbsp;</div>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method-summary">
<h2>Method Summary</h2>
<div id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button id="method-summary-table-tab0" role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table', 3)" class="active-table-tab">All Methods</button><button id="method-summary-table-tab2" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab2', 3)" class="table-tab">Instance Methods</button><button id="method-summary-table-tab4" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab4', 3)" class="table-tab">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<div class="summary-table three-column-summary" aria-labelledby="method-summary-table-tab0">
<div class="table-header col-first">Modifier and Type</div>
<div class="table-header col-second">Method</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#clone()" class="member-name-link">clone</a>()</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code>double[]</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#getMarginalDistribution(int)" class="member-name-link">getMarginalDistribution</a><wbr>(int&nbsp;gene)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">&nbsp;</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code>private double</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#marginalDist(int,int,int%5B%5D%5B%5D)" class="member-name-link">marginalDist</a><wbr>(int&nbsp;i,
 int&nbsp;k,
 int[][]&nbsp;arz)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">&nbsp;</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="../../Individual.html" title="class in ec">Individual</a></code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#newIndividual(ec.EvolutionState,int)" class="member-name-link">newIndividual</a><wbr>(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 int&nbsp;thread)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">Provides a brand-new individual to fill in a population.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code>void</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#setup(ec.EvolutionState,ec.util.Parameter)" class="member-name-link">setup</a><wbr>(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code>void</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#updateDistribution(ec.EvolutionState,ec.Subpopulation)" class="member-name-link">updateDistribution</a><wbr>(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../Subpopulation.html" title="class in ec">Subpopulation</a>&nbsp;subpop)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">Revises the PBIL distribution to reflect the current fitness results in the provided subpopulation.</div>
</div>
</div>
</div>
</div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-ec.vector.IntegerVectorSpecies">Methods inherited from class&nbsp;ec.vector.<a href="../../vector/IntegerVectorSpecies.html" title="class in ec.vector">IntegerVectorSpecies</a></h3>
<code><a href="../../vector/IntegerVectorSpecies.html#inNumericalTypeRange(double)">inNumericalTypeRange</a>, <a href="../../vector/IntegerVectorSpecies.html#inNumericalTypeRange(long)">inNumericalTypeRange</a>, <a href="../../vector/IntegerVectorSpecies.html#loadParametersForGene(ec.EvolutionState,int,ec.util.Parameter,ec.util.Parameter,java.lang.String)">loadParametersForGene</a>, <a href="../../vector/IntegerVectorSpecies.html#maxGene(int)">maxGene</a>, <a href="../../vector/IntegerVectorSpecies.html#minGene(int)">minGene</a>, <a href="../../vector/IntegerVectorSpecies.html#mutationIsBounded(int)">mutationIsBounded</a>, <a href="../../vector/IntegerVectorSpecies.html#mutationType(int)">mutationType</a>, <a href="../../vector/IntegerVectorSpecies.html#randomWalkProbability(int)">randomWalkProbability</a></code></div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-ec.vector.VectorSpecies">Methods inherited from class&nbsp;ec.vector.<a href="../../vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></h3>
<code><a href="../../vector/VectorSpecies.html#contains(boolean%5B%5D,boolean)">contains</a>, <a href="../../vector/VectorSpecies.html#contains(double%5B%5D,double)">contains</a>, <a href="../../vector/VectorSpecies.html#contains(int%5B%5D,int)">contains</a>, <a href="../../vector/VectorSpecies.html#contains(long%5B%5D,long)">contains</a>, <a href="../../vector/VectorSpecies.html#defaultBase()">defaultBase</a>, <a href="../../vector/VectorSpecies.html#duplicateRetries(int)">duplicateRetries</a>, <a href="../../vector/VectorSpecies.html#fill(boolean%5B%5D,boolean)">fill</a>, <a href="../../vector/VectorSpecies.html#fill(double%5B%5D,double)">fill</a>, <a href="../../vector/VectorSpecies.html#fill(int%5B%5D,int)">fill</a>, <a href="../../vector/VectorSpecies.html#fill(long%5B%5D,long)">fill</a>, <a href="../../vector/VectorSpecies.html#initializeGenomeSegmentsByEndIndices(ec.EvolutionState,ec.util.Parameter,ec.util.Parameter,int)">initializeGenomeSegmentsByEndIndices</a>, <a href="../../vector/VectorSpecies.html#initializeGenomeSegmentsByStartIndices(ec.EvolutionState,ec.util.Parameter,ec.util.Parameter,int)">initializeGenomeSegmentsByStartIndices</a>, <a href="../../vector/VectorSpecies.html#mutationProbability(int)">mutationProbability</a>, <a href="../../vector/VectorSpecies.html#setupGenome(ec.EvolutionState,ec.util.Parameter)">setupGenome</a></code></div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-ec.Species">Methods inherited from class&nbsp;ec.<a href="../../Species.html" title="class in ec">Species</a></h3>
<code><a href="../../Species.html#buildMisc(ec.EvolutionState,int,int)">buildMisc</a>, <a href="../../Species.html#newIndividual(ec.EvolutionState,java.io.DataInput)">newIndividual</a>, <a href="../../Species.html#newIndividual(ec.EvolutionState,java.io.LineNumberReader)">newIndividual</a>, <a href="../../Species.html#updateIndividual(ec.EvolutionState,ec.Individual)">updateIndividual</a>, <a href="../../Species.html#updateSubpopulation(ec.EvolutionState,ec.Subpopulation)">updateSubpopulation</a></code></div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-java.lang.Object">Methods inherited from class&nbsp;java.lang.<a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></h3>
<code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#equals(java.lang.Object)" title="class or interface in java.lang" class="external-link">equals</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#finalize()" title="class or interface in java.lang" class="external-link">finalize</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#getClass()" title="class or interface in java.lang" class="external-link">getClass</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#hashCode()" title="class or interface in java.lang" class="external-link">hashCode</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#notify()" title="class or interface in java.lang" class="external-link">notify</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#notifyAll()" title="class or interface in java.lang" class="external-link">notifyAll</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#toString()" title="class or interface in java.lang" class="external-link">toString</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#wait()" title="class or interface in java.lang" class="external-link">wait</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#wait(long)" title="class or interface in java.lang" class="external-link">wait</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#wait(long,int)" title="class or interface in java.lang" class="external-link">wait</a></code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field-detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="P_PBIL_SPECIES">
<h3>P_PBIL_SPECIES</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_PBIL_SPECIES</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.pbil.PBILSpecies.P_PBIL_SPECIES">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_ALPHA">
<h3>P_ALPHA</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_ALPHA</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.pbil.PBILSpecies.P_ALPHA">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_B">
<h3>P_B</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_B</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.pbil.PBILSpecies.P_B">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="alpha">
<h3>alpha</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">alpha</span></div>
</section>
</li>
<li>
<section class="detail" id="b">
<h3>b</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">b</span></div>
</section>
</li>
<li>
<section class="detail" id="distributions">
<h3>distributions</h3>
<div class="member-signature"><span class="modifiers">private</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/util/List.html" title="class or interface in java.util" class="external-link">List</a>&lt;double[]&gt;</span>&nbsp;<span class="element-name">distributions</span></div>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor-detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;()">
<h3>PBILSpecies</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="element-name">PBILSpecies</span>()</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method-detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="getMarginalDistribution(int)">
<h3>getMarginalDistribution</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double[]</span>&nbsp;<span class="element-name">getMarginalDistribution</span><wbr><span class="parameters">(int&nbsp;gene)</span></div>
</section>
</li>
<li>
<section class="detail" id="setup(ec.EvolutionState,ec.util.Parameter)">
<h3>setup</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name">setup</span><wbr><span class="parameters">(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</span></div>
<div class="block"><span class="description-from-type-label">Description copied from class:&nbsp;<code><a href="../../Species.html#setup(ec.EvolutionState,ec.util.Parameter)">Species</a></code></span></div>
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.  
        If your individual prototype might need to know special things about the species (like parameters stored in it),
        then when you override this setup method, you'll need to set those parameters BEFORE you call super.setup(...),
        because the setup(...) code in Species sets up the prototype.</div>
<dl class="notes">
<dt>Specified by:</dt>
<dd><code><a href="../../Prototype.html#setup(ec.EvolutionState,ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt>Specified by:</dt>
<dd><code><a href="../../Setup.html#setup(ec.EvolutionState,ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../Setup.html" title="interface in ec">Setup</a></code></dd>
<dt>Overrides:</dt>
<dd><code><a href="../../vector/IntegerVectorSpecies.html#setup(ec.EvolutionState,ec.util.Parameter)">setup</a></code>&nbsp;in class&nbsp;<code><a href="../../vector/IntegerVectorSpecies.html" title="class in ec.vector">IntegerVectorSpecies</a></code></dd>
<dt>See Also:</dt>
<dd>
<ul class="see-list-long">
<li><a href="../../Prototype.html#setup(ec.EvolutionState,ec.util.Parameter)"><code>Prototype.setup(EvolutionState,Parameter)</code></a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="clone()">
<h3>clone</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></span>&nbsp;<span class="element-name">clone</span>()</div>
<div class="block"><span class="description-from-type-label">Description copied from interface:&nbsp;<code><a href="../../Prototype.html#clone()">Prototype</a></code></span></div>
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.

        <p>Typically this should be a full "deep" clone.
        However, you may share certain elements with other objects
        rather than clone hem, depending on the situation:

        <p>
        <ul>
        <li>If you hold objects which are shared with other instances,
        don't clone them.
        <li>If you hold objects which must be unique, clone them.
        <li>If you hold objects which were given to you as a gesture
        of kindness, and aren't owned by you, you probably shouldn't clone
        them.
        <li> DON'T attempt to clone: Singletons, Cliques, or Populations, or Subpopulation.
        <li>Arrays are not cloned automatically; you may need to
        clone an array if you're not sharing it with other instances.
        Arrays have the nice feature of being copyable by calling clone()
        on them.
        </ul>

        <p><b>Implementations.</b>

        <ul>
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are abstract, then you should not declare clone().
        
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are <b>not</b> abstract, then you should implement
        it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         return super.clone();
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     }
        </pre></tt>
        
        <li>If no ancestor of yours implements clone(), but you
        need to deep-clone some things, then you should implement it
        as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         MyObject myobj = (MyObject) (super.clone());

         // put your deep-cloning code here...
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     return myobj;
     } 
        </pre></tt>

        <li>If an ancestor has implemented clone(), and you also need
        to deep clone some things, then you should implement it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     { 
     MyObject myobj = (MyObject) (super.clone());

     // put your deep-cloning code here...

     return myobj;
     } 
        </pre></tt>
        </ul></div>
<dl class="notes">
<dt>Specified by:</dt>
<dd><code><a href="../../Prototype.html#clone()">clone</a></code>&nbsp;in interface&nbsp;<code><a href="../../Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt>Overrides:</dt>
<dd><code><a href="../../Species.html#clone()">clone</a></code>&nbsp;in class&nbsp;<code><a href="../../Species.html" title="class in ec">Species</a></code></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="newIndividual(ec.EvolutionState,int)">
<h3>newIndividual</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type"><a href="../../Individual.html" title="class in ec">Individual</a></span>&nbsp;<span class="element-name">newIndividual</span><wbr><span class="parameters">(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 int&nbsp;thread)</span></div>
<div class="block"><span class="description-from-type-label">Description copied from class:&nbsp;<code><a href="../../Species.html#newIndividual(ec.EvolutionState,int)">Species</a></code></span></div>
<div class="block">Provides a brand-new individual to fill in a population.  The default form
        simply calls clone(), creates a fitness, sets evaluated to false, and sets
        the species.  If you need to make a more custom genotype (as is the case
        for GPSpecies, which requires a light rather than deep clone), 
        you will need to override this method as you see fit.</div>
<dl class="notes">
<dt>Overrides:</dt>
<dd><code><a href="../../vector/VectorSpecies.html#newIndividual(ec.EvolutionState,int)">newIndividual</a></code>&nbsp;in class&nbsp;<code><a href="../../vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></code></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="updateDistribution(ec.EvolutionState,ec.Subpopulation)">
<h3>updateDistribution</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name">updateDistribution</span><wbr><span class="parameters">(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../Subpopulation.html" title="class in ec">Subpopulation</a>&nbsp;subpop)</span></div>
<div class="block">Revises the PBIL distribution to reflect the current fitness results in the provided subpopulation.</div>
</section>
</li>
<li>
<section class="detail" id="marginalDist(int,int,int[][])">
<h3>marginalDist</h3>
<div class="member-signature"><span class="modifiers">private</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">marginalDist</span><wbr><span class="parameters">(int&nbsp;i,
 int&nbsp;k,
 int[][]&nbsp;arz)</span></div>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<hr>
<p class="legal-copy"><small>Copyright &#169; 2023. All rights reserved.</small></p>
</footer>
</div>
</div>
</body>
</html>
