<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>ByteVectorIndividual.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">ecj</a> &gt; <a href="index.source.html" class="el_package">ec.vector</a> &gt; <span class="el_source">ByteVectorIndividual.java</span></div><h1>ByteVectorIndividual.java</h1><pre class="source lang-java linenums">/*
  Copyright 2006 by Sean Luke and George Mason University
  Licensed under the Academic Free License version 3.0
  See the file &quot;LICENSE&quot; for more information
*/


package ec.vector;

import ec.*;
import ec.util.*;
import java.io.*;

/*
 * ByteVectorIndividual.java
 * Created: Tue Mar 13 15:03:12 EST 2001
 */

/**
 * ByteVectorIndividual is a VectorIndividual whose genome is an array of bytes.
 * Gene values may range from species.mingene(x) to species.maxgene(x), inclusive.
 * The default mutation method randomizes genes to new values in this range,
 * with &lt;tt&gt;species.mutationProbability&lt;/tt&gt;.

 * &lt;P&gt;&lt;b&gt;From ec.Individual:&lt;/b&gt;  
 *
 * &lt;p&gt;In addition to serialization for checkpointing, Individuals may read and write themselves to streams in three ways.
 *
 * &lt;ul&gt;
 * &lt;li&gt;&lt;b&gt;writeIndividual(...,DataOutput)/readIndividual(...,DataInput)&lt;/b&gt;&amp;nbsp;&amp;nbsp;&amp;nbsp;This method
 * transmits or receives an individual in binary.  It is the most efficient approach to sending
 * individuals over networks, etc.  These methods write the evaluated flag and the fitness, then
 * call &lt;b&gt;readGenotype/writeGenotype&lt;/b&gt;, which you must implement to write those parts of your 
 * Individual special to your functions-- the default versions of readGenotype/writeGenotype throw errors.
 * You don't need to implement them if you don't plan on using read/writeIndividual.
 *
 * &lt;li&gt;&lt;b&gt;printIndividual(...,PrintWriter)/readIndividual(...,LineNumberReader)&lt;/b&gt;&amp;nbsp;&amp;nbsp;&amp;nbsp;This
 * approach transmits or receives an indivdual in text encoded such that the individual is largely readable
 * by humans but can be read back in 100% by ECJ as well.  To do this, these methods will encode numbers
 * using the &lt;tt&gt;ec.util.Code&lt;/tt&gt; class.  These methods are mostly used to write out populations to
 * files for inspection, slight modification, then reading back in later on.  &lt;b&gt;readIndividual&lt;/b&gt; reads
 * in the fitness and the evaluation flag, then calls &lt;b&gt;parseGenotype&lt;/b&gt; to read in the remaining individual.
 * You are responsible for implementing parseGenotype: the Code class is there to help you.
 * &lt;b&gt;printIndividual&lt;/b&gt; writes out the fitness and evaluation flag, then calls &lt;b&gt;genotypeToString&lt;/b&gt; 
 * and printlns the resultant string. You are responsible for implementing the genotypeToString method in such
 * a way that parseGenotype can read back in the individual println'd with genotypeToString.  The default form
 * of genotypeToString simply calls &lt;b&gt;toString&lt;/b&gt;, which you may override instead if you like.  The default
 * form of &lt;b&gt;parseGenotype&lt;/b&gt; throws an error.  You are not required to implement these methods, but without
 * them you will not be able to write individuals to files in a simultaneously computer- and human-readable fashion.
 *
 * &lt;li&gt;&lt;b&gt;printIndividualForHumans(...,PrintWriter)&lt;/b&gt;&amp;nbsp;&amp;nbsp;&amp;nbsp;This
 * approach prints an individual in a fashion intended for human consumption only.
 * &lt;b&gt;printIndividualForHumans&lt;/b&gt; writes out the fitness and evaluation flag, then calls &lt;b&gt;genotypeToStringForHumans&lt;/b&gt; 
 * and printlns the resultant string. You are responsible for implementing the genotypeToStringForHumans method.
 * The default form of genotypeToStringForHumans simply calls &lt;b&gt;toString&lt;/b&gt;, which you may override instead if you like
 * (though note that genotypeToString's default also calls toString).  You should handle one of these methods properly
 * to ensure individuals can be printed by ECJ.
 * &lt;/ul&gt;

 * &lt;p&gt;In general, the various readers and writers do three things: they tell the Fitness to read/write itself,
 * they read/write the evaluated flag, and they read/write the gene array.  If you add instance variables to
 * a VectorIndividual or subclass, you'll need to read/write those variables as well.
 &lt;p&gt;&lt;b&gt;Default Base&lt;/b&gt;&lt;br&gt;
 vector.byte-vect-ind

 * @author Liviu Panait
 * @author Sean Luke
 * @version 1.0
 */

<span class="nc" id="L71">public class ByteVectorIndividual extends VectorIndividual</span>
    {
    public static final String P_BYTEVECTORINDIVIDUAL = &quot;byte-vect-ind&quot;;
    public byte[] genome;
    
    public Parameter defaultBase()
        {
<span class="nc" id="L78">        return VectorDefaults.base().push(P_BYTEVECTORINDIVIDUAL);</span>
        }

    public Object clone()
        {
<span class="nc" id="L83">        ByteVectorIndividual myobj = (ByteVectorIndividual) (super.clone());</span>

        // must clone the genome
<span class="nc" id="L86">        myobj.genome = (byte[])(genome.clone());</span>
        
<span class="nc" id="L88">        return myobj;</span>
        } 

    public void setup(final EvolutionState state, final Parameter base)
        {
<span class="nc" id="L93">        super.setup(state,base);  // actually unnecessary (Individual.setup() is empty)</span>

        // since IntegerVectorSpecies set its constraint values BEFORE it called
        // super.setup(...) [which in turn called our setup(...)], we know that
        // stuff like genomeSize has already been set...
        
<span class="nc" id="L99">        Parameter def = defaultBase();</span>
        
<span class="nc bnc" id="L101" title="All 2 branches missed.">        if (!(species instanceof IntegerVectorSpecies)) </span>
<span class="nc" id="L102">            state.output.fatal(&quot;ByteVectorIndividual requires an IntegerVectorSpecies&quot;, base, def);</span>
<span class="nc" id="L103">        IntegerVectorSpecies s = (IntegerVectorSpecies) species;</span>
        
<span class="nc" id="L105">        genome = new byte[s.genomeSize];</span>
<span class="nc" id="L106">        }</span>
        
    public void defaultCrossover(EvolutionState state, int thread, VectorIndividual ind)
        {
<span class="nc" id="L110">        IntegerVectorSpecies s = (IntegerVectorSpecies) species;</span>
<span class="nc" id="L111">        ByteVectorIndividual i = (ByteVectorIndividual) ind;</span>
        byte tmp;
        int point;
        
<span class="nc" id="L115">        int len = Math.min(genome.length, i.genome.length);</span>
<span class="nc bnc" id="L116" title="All 4 branches missed.">        if (len != genome.length || len != i.genome.length)</span>
<span class="nc" id="L117">            state.output.warnOnce(&quot;Genome lengths are not the same.  Vector crossover will only be done in overlapping region.&quot;);</span>

<span class="nc bnc" id="L119" title="All 8 branches missed.">        switch(s.crossoverType)</span>
            {
            case VectorSpecies.C_ONE_POINT:
                //                point = state.random[thread].nextInt((len / s.chunksize)+1);
                // we want to go from 0 ... len-1 
                // so that there is only ONE case of NO-OP crossover, not TWO
<span class="nc" id="L125">                point = state.random[thread].nextInt((len / s.chunksize));</span>
<span class="nc bnc" id="L126" title="All 2 branches missed.">                for(int x=0;x&lt;point*s.chunksize;x++)</span>
                    { 
<span class="nc" id="L128">                    tmp = i.genome[x];</span>
<span class="nc" id="L129">                    i.genome[x] = genome[x]; </span>
<span class="nc" id="L130">                    genome[x] = tmp; </span>
                    }
<span class="nc" id="L132">                break;</span>
            case VectorSpecies.C_ONE_POINT_NO_NOP:
<span class="nc" id="L134">                point = state.random[thread].nextInt((len / s.chunksize) - 1) + 1;  // so it goes from 1 .. len-1</span>
<span class="nc bnc" id="L135" title="All 2 branches missed.">                for(int x=0;x&lt;point*s.chunksize;x++)</span>
                    { 
<span class="nc" id="L137">                    tmp = i.genome[x];</span>
<span class="nc" id="L138">                    i.genome[x] = genome[x]; </span>
<span class="nc" id="L139">                    genome[x] = tmp; </span>
                    }
<span class="nc" id="L141">                break;</span>
            case VectorSpecies.C_TWO_POINT: 
                {
                //                int point0 = state.random[thread].nextInt((len / s.chunksize)+1);
                //                point = state.random[thread].nextInt((len / s.chunksize)+1);
                // we want to go from 0 to len-1
                // so that the only NO-OP crossover possible is point == point0
                // example; len = 4
                // possibilities: a=0 b=0       NOP                             [0123]
                //                                a=0 b=1       swap 0                  [for 1, 2, 3]
                //                                a=0 b=2       swap 0, 1               [for 2, 3]
                //                                a=0 b=3       swap 0, 1, 2    [for 3]
                //                                a=1 b=1       NOP                             [1230]
                //                                a=1 b=2       swap 1                  [for 2, 3, 0]
                //                                a=1 b=3       swap 1, 2               [for 3, 0]
                //                                a=2 b=2       NOP                             [2301]
                //                                a=2 b=3       swap 2                  [for 3, 0, 1]
                //                                a=3 b=3   NOP                         [3012]
                // All intervals: 0, 01, 012, 0123, 1, 12, 123, 1230, 2, 23, 230, 2301, 3, 30, 301, 3012
<span class="nc" id="L160">                point = state.random[thread].nextInt((len / s.chunksize));</span>
<span class="nc" id="L161">                int point0 = state.random[thread].nextInt((len / s.chunksize));</span>
<span class="nc bnc" id="L162" title="All 2 branches missed.">                if (point0 &gt; point) { int p = point0; point0 = point; point = p; }</span>
<span class="nc bnc" id="L163" title="All 2 branches missed.">                for(int x=point0*s.chunksize;x&lt;point*s.chunksize;x++)</span>
                    {
<span class="nc" id="L165">                    tmp = i.genome[x];</span>
<span class="nc" id="L166">                    i.genome[x] = genome[x];</span>
<span class="nc" id="L167">                    genome[x] = tmp;</span>
                    }
                }
<span class="nc" id="L170">            break;</span>
            case VectorSpecies.C_TWO_POINT_NO_NOP: 
                {
<span class="nc" id="L173">                point = state.random[thread].nextInt((len / s.chunksize));</span>
<span class="nc" id="L174">                int point0 = 0;</span>
<span class="nc" id="L175">                do { point0 = state.random[thread].nextInt((len / s.chunksize)); }</span>
<span class="nc bnc" id="L176" title="All 2 branches missed.">                while (point0 == point);  // NOP</span>
<span class="nc bnc" id="L177" title="All 2 branches missed.">                if (point0 &gt; point) { int p = point0; point0 = point; point = p; }</span>
<span class="nc bnc" id="L178" title="All 2 branches missed.">                for(int x=point0*s.chunksize;x&lt;point*s.chunksize;x++)</span>
                    {
<span class="nc" id="L180">                    tmp = i.genome[x];</span>
<span class="nc" id="L181">                    i.genome[x] = genome[x];</span>
<span class="nc" id="L182">                    genome[x] = tmp;</span>
                    }
                }
<span class="nc" id="L185">            break;</span>
            case VectorSpecies.C_ANY_POINT:
<span class="nc bnc" id="L187" title="All 2 branches missed.">                for(int x=0;x&lt;len/s.chunksize;x++) </span>
<span class="nc bnc" id="L188" title="All 2 branches missed.">                    if (state.random[thread].nextBoolean(s.crossoverProbability))</span>
<span class="nc bnc" id="L189" title="All 2 branches missed.">                        for(int y=x*s.chunksize;y&lt;(x+1)*s.chunksize;y++)</span>
                            {
<span class="nc" id="L191">                            tmp = i.genome[y];</span>
<span class="nc" id="L192">                            i.genome[y] = genome[y];</span>
<span class="nc" id="L193">                            genome[y] = tmp;</span>
                            }
<span class="nc" id="L195">                break;</span>
            case VectorSpecies.C_LINE_RECOMB:
                {
<span class="nc" id="L198">                double alpha = state.random[thread].nextDouble() * (1 + 2*s.lineDistance) - s.lineDistance;</span>
<span class="nc" id="L199">                double beta = state.random[thread].nextDouble() * (1 + 2*s.lineDistance) - s.lineDistance;</span>
                long t,u;
                long min, max;
<span class="nc bnc" id="L202" title="All 2 branches missed.">                for (int x = 0; x &lt; len; x++)</span>
                    {
<span class="nc" id="L204">                    min = s.minGene(x);</span>
<span class="nc" id="L205">                    max = s.maxGene(x);</span>
<span class="nc" id="L206">                    t = (long) Math.floor(alpha * genome[x] + (1 - alpha) * i.genome[x] + 0.5);</span>
<span class="nc" id="L207">                    u = (long) Math.floor(beta * i.genome[x] + (1 - beta) * genome[x] + 0.5);</span>
<span class="nc bnc" id="L208" title="All 8 branches missed.">                    if (!(t &lt; min || t &gt; max || u &lt; min || u &gt; max))</span>
                        {
<span class="nc" id="L210">                        genome[x] = (byte) t;</span>
<span class="nc" id="L211">                        i.genome[x] = (byte) u; </span>
                        }
                    }
                }
<span class="nc" id="L215">            break;</span>
            case VectorSpecies.C_INTERMED_RECOMB:
                {
                long t,u;
                long min, max;
<span class="nc bnc" id="L220" title="All 2 branches missed.">                for (int x = 0; x &lt; len; x++)</span>
                    {
                    do
                        {
<span class="nc" id="L224">                        double alpha = state.random[thread].nextDouble() * (1 + 2*s.lineDistance) - s.lineDistance;</span>
<span class="nc" id="L225">                        double beta = state.random[thread].nextDouble() * (1 + 2*s.lineDistance) - s.lineDistance;</span>
<span class="nc" id="L226">                        min = s.minGene(x);</span>
<span class="nc" id="L227">                        max = s.maxGene(x);</span>
<span class="nc" id="L228">                        t = (long) Math.floor(alpha * genome[x] + (1 - alpha) * i.genome[x] + 0.5);</span>
<span class="nc" id="L229">                        u = (long) Math.floor(beta * i.genome[x] + (1 - beta) * genome[x] + 0.5);</span>
<span class="nc bnc" id="L230" title="All 8 branches missed.">                        } while (t &lt; min || t &gt; max || u &lt; min || u &gt; max);</span>
<span class="nc" id="L231">                    genome[x] = (byte) t;</span>
<span class="nc" id="L232">                    i.genome[x] = (byte) u; </span>
                    }
                }
            break;
            }
<span class="nc" id="L237">        }</span>

    /** Splits the genome into n pieces, according to points, which *must* be sorted. 
        pieces.length must be 1 + points.length */
    public void split(int[] points, Object[] pieces)
        {
        int point0, point1;
<span class="nc" id="L244">        point0 = 0; point1 = points[0];</span>
<span class="nc bnc" id="L245" title="All 2 branches missed.">        for(int x=0;x&lt;pieces.length;x++)</span>
            {
<span class="nc" id="L247">            pieces[x] = new byte[point1-point0];</span>
<span class="nc" id="L248">            System.arraycopy(genome,point0,pieces[x],0,point1-point0);</span>
<span class="nc" id="L249">            point0 = point1;</span>
<span class="nc bnc" id="L250" title="All 2 branches missed.">            if (x &gt;=pieces.length-2)</span>
<span class="nc" id="L251">                point1 = genome.length;</span>
<span class="nc" id="L252">            else point1 = points[x+1];</span>
            }
<span class="nc" id="L254">        }</span>
    
    /** Joins the n pieces and sets the genome to their concatenation.*/
    public void join(Object[] pieces)
        {
<span class="nc" id="L259">        int sum=0;</span>
<span class="nc bnc" id="L260" title="All 2 branches missed.">        for(int x=0;x&lt;pieces.length;x++)</span>
<span class="nc" id="L261">            sum += ((byte[])(pieces[x])).length;</span>
        
<span class="nc" id="L263">        int runningsum = 0;</span>
<span class="nc" id="L264">        byte[] newgenome = new byte[sum];</span>
<span class="nc bnc" id="L265" title="All 2 branches missed.">        for(int x=0;x&lt;pieces.length;x++)</span>
            {
<span class="nc" id="L267">            System.arraycopy(pieces[x], 0, newgenome, runningsum, ((byte[])(pieces[x])).length);</span>
<span class="nc" id="L268">            runningsum += ((byte[])(pieces[x])).length;</span>
            }
        // set genome
<span class="nc" id="L271">        genome = newgenome;</span>
<span class="nc" id="L272">        }</span>


    /** Returns a random value from between min and max inclusive.  This method handles
        overflows that complicate this computation.  Does NOT check that
        min is less than or equal to max.  You must check this yourself. */
    public byte randomValueFromClosedInterval(byte min, byte max, MersenneTwisterFast random)
        {
<span class="nc bnc" id="L280" title="All 2 branches missed.">        if (max - min &lt; 0) // we had an overflow</span>
            {
<span class="nc" id="L282">            byte l = 0;</span>
<span class="nc" id="L283">            do l = (byte)random.nextInt();</span>
<span class="nc bnc" id="L284" title="All 4 branches missed.">            while(l &lt; min || l &gt; max);</span>
<span class="nc" id="L285">            return l;</span>
            }
<span class="nc" id="L287">        else return (byte)(min + random.nextInt(max - min + 1));</span>
        }

    /** Destructively mutates the individual in some default manner.  The default form
        simply randomizes genes to a uniform distribution from the min and max of the gene values. */
    public void defaultMutate(EvolutionState state, int thread)
        {
<span class="nc" id="L294">        IntegerVectorSpecies s = (IntegerVectorSpecies) species;</span>
<span class="nc bnc" id="L295" title="All 2 branches missed.">        for(int x = 0; x &lt; genome.length; x++)</span>
<span class="nc bnc" id="L296" title="All 2 branches missed.">            if (state.random[thread].nextBoolean(s.mutationProbability(x)))</span>
                {
<span class="nc" id="L298">                byte old = genome[x];</span>
<span class="nc bnc" id="L299" title="All 2 branches missed.">                for(int retries = 0; retries &lt; s.duplicateRetries(x) + 1; retries++)</span>
                    {
<span class="nc bnc" id="L301" title="All 3 branches missed.">                    switch(s.mutationType(x))</span>
                        {
                        case IntegerVectorSpecies.C_RESET_MUTATION:
<span class="nc" id="L304">                            genome[x] = (byte)randomValueFromClosedInterval((byte)s.minGene(x), (byte)s.maxGene(x), state.random[thread]);</span>
<span class="nc" id="L305">                            break;</span>
                        case IntegerVectorSpecies.C_RANDOM_WALK_MUTATION:
<span class="nc" id="L307">                            int min = (int)s.minGene(x);</span>
<span class="nc" id="L308">                            int max = (int)s.maxGene(x);</span>
<span class="nc bnc" id="L309" title="All 2 branches missed.">                            if (!s.mutationIsBounded(x))</span>
                                {
                                // okay, technically these are still bounds, but we can't go beyond this without weird things happening
<span class="nc" id="L312">                                max = Byte.MAX_VALUE;</span>
<span class="nc" id="L313">                                min = Byte.MIN_VALUE;</span>
                                }
                            do
                                {
<span class="nc bnc" id="L317" title="All 2 branches missed.">                                int n = (int)(state.random[thread].nextBoolean() ? 1 : -1);</span>
<span class="nc" id="L318">                                int g = genome[x];</span>
<span class="nc bnc" id="L319" title="All 8 branches missed.">                                if ((n == 1 &amp;&amp; g &lt; max) ||</span>
                                    (n == -1 &amp;&amp; g &gt; min))
<span class="nc" id="L321">                                    genome[x] = (byte)(g + n);</span>
<span class="nc bnc" id="L322" title="All 8 branches missed.">                                else if ((n == -1 &amp;&amp; g &lt; max) ||</span>
                                    (n == 1 &amp;&amp; g &gt; min))
<span class="nc" id="L324">                                    genome[x] = (byte)(g - n);     </span>
                                }
<span class="nc bnc" id="L326" title="All 2 branches missed.">                            while (state.random[thread].nextBoolean(s.randomWalkProbability(x)));</span>
<span class="nc" id="L327">                            break;</span>
                        default:
<span class="nc" id="L329">                            state.output.fatal(&quot;In ByteVectorIndividual.defaultMutate, default case occurred when it shouldn't have&quot;);</span>
                            break;
                        }
<span class="nc bnc" id="L332" title="All 2 branches missed.">                    if (genome[x] != old) break;</span>
                    // else genome[x] = old;  // try again
                    }
                }
<span class="nc" id="L336">        }</span>
        
    
    /** Initializes the individual by randomly choosing Integers uniformly from mingene to maxgene. */
    // notice that we bump to longs to avoid overflow errors
    public void reset(EvolutionState state, int thread)
        {
<span class="nc" id="L343">        IntegerVectorSpecies s = (IntegerVectorSpecies) species;</span>
<span class="nc bnc" id="L344" title="All 2 branches missed.">        for(int x=0;x&lt;genome.length;x++)</span>
<span class="nc" id="L345">            genome[x] = (byte)randomValueFromClosedInterval((byte)s.minGene(x), (byte)s.maxGene(x), state.random[thread]);</span>
<span class="nc" id="L346">        }</span>


    public int hashCode()
        {
        // stolen from GPIndividual.  It's a decent algorithm.
<span class="nc" id="L352">        int hash = this.getClass().hashCode();</span>

<span class="nc" id="L354">        hash = ( hash &lt;&lt; 1 | hash &gt;&gt;&gt; 31 );</span>
<span class="nc bnc" id="L355" title="All 2 branches missed.">        for(int x=0;x&lt;genome.length;x++)</span>
<span class="nc" id="L356">            hash = ( hash &lt;&lt; 1 | hash &gt;&gt;&gt; 31 ) ^ genome[x];</span>

<span class="nc" id="L358">        return hash;</span>
        }

    public String genotypeToStringForHumans()
        {
<span class="nc" id="L363">        StringBuilder s = new StringBuilder();</span>
<span class="nc bnc" id="L364" title="All 2 branches missed.">        for( int i = 0 ; i &lt; genome.length ; i++ )</span>
<span class="nc bnc" id="L365" title="All 2 branches missed.">            { if (i &gt; 0) s.append(&quot; &quot;); s.append(genome[i]); }</span>
<span class="nc" id="L366">        return s.toString();</span>
        }
        
    public String genotypeToString()
        {
<span class="nc" id="L371">        StringBuilder s = new StringBuilder();</span>
<span class="nc" id="L372">        s.append( Code.encode( genome.length ) );</span>
<span class="nc bnc" id="L373" title="All 2 branches missed.">        for( int i = 0 ; i &lt; genome.length ; i++ )</span>
<span class="nc" id="L374">            s.append( Code.encode( genome[i] ) );</span>
<span class="nc" id="L375">        return s.toString();</span>
        }
                
    protected void parseGenotype(final EvolutionState state,
        final LineNumberReader reader) throws IOException
        {
        // read in the next line.  The first item is the number of genes
<span class="nc" id="L382">        String s = reader.readLine();</span>
<span class="nc" id="L383">        DecodeReturn d = new DecodeReturn(s);</span>
<span class="nc" id="L384">        Code.decode( d );</span>
<span class="nc bnc" id="L385" title="All 2 branches missed.">        if (d.type != DecodeReturn.T_INTEGER)  // uh oh</span>
<span class="nc" id="L386">            state.output.fatal(&quot;Individual with genome:\n&quot; + s + &quot;\n... does not have an integer at the beginning indicating the genome count.&quot;);</span>
<span class="nc" id="L387">        int lll = (int)(d.l);</span>

<span class="nc" id="L389">        genome = new byte[ lll ];</span>

        // read in the genes
<span class="nc bnc" id="L392" title="All 2 branches missed.">        for( int i = 0 ; i &lt; genome.length ; i++ )</span>
            {
<span class="nc" id="L394">            Code.decode( d );</span>
<span class="nc" id="L395">            genome[i] = (byte)(d.l);</span>
            }
<span class="nc" id="L397">        }</span>

    public boolean equals(Object ind)
        {
<span class="nc bnc" id="L401" title="All 2 branches missed.">        if (ind == null) return false;</span>
<span class="nc bnc" id="L402" title="All 2 branches missed.">        if (!(this.getClass().equals(ind.getClass()))) return false; // SimpleRuleIndividuals are special.</span>
<span class="nc" id="L403">        ByteVectorIndividual i = (ByteVectorIndividual)ind;</span>
<span class="nc bnc" id="L404" title="All 2 branches missed.">        if( genome.length != i.genome.length )</span>
<span class="nc" id="L405">            return false;</span>
<span class="nc bnc" id="L406" title="All 2 branches missed.">        for( int j = 0 ; j &lt; genome.length ; j++ )</span>
<span class="nc bnc" id="L407" title="All 2 branches missed.">            if( genome[j] != i.genome[j] )</span>
<span class="nc" id="L408">                return false;</span>
<span class="nc" id="L409">        return true;</span>
        }

    public Object getGenome()
<span class="nc" id="L413">        { return genome; }</span>
    public void setGenome(Object gen)
<span class="nc" id="L415">        { genome = (byte[]) gen; }</span>
    public int genomeLength()
<span class="nc" id="L417">        { return genome.length; }</span>

    public void writeGenotype(final EvolutionState state,
        final DataOutput dataOutput) throws IOException
        {
<span class="nc" id="L422">        dataOutput.writeInt(genome.length);</span>
<span class="nc bnc" id="L423" title="All 2 branches missed.">        for(int x=0;x&lt;genome.length;x++)</span>
<span class="nc" id="L424">            dataOutput.writeByte(genome[x]);</span>
<span class="nc" id="L425">        }</span>

    public void readGenotype(final EvolutionState state,
        final DataInput dataInput) throws IOException
        {
<span class="nc" id="L430">        int len = dataInput.readInt();</span>
<span class="nc bnc" id="L431" title="All 4 branches missed.">        if (genome==null || genome.length != len)</span>
<span class="nc" id="L432">            genome = new byte[len];</span>
<span class="nc bnc" id="L433" title="All 2 branches missed.">        for(int x=0;x&lt;genome.length;x++)</span>
<span class="nc" id="L434">            genome[x] = dataInput.readByte();</span>
<span class="nc" id="L435">        }</span>

    /** Clips each gene value to be within its specified [min,max] range. */
    public void clamp() 
        {
<span class="nc" id="L440">        IntegerVectorSpecies _species = (IntegerVectorSpecies)species;</span>
<span class="nc bnc" id="L441" title="All 2 branches missed.">        for (int i = 0; i &lt; genomeLength(); i++)</span>
            {
<span class="nc" id="L443">            byte minGene = (byte)_species.minGene(i);</span>
<span class="nc bnc" id="L444" title="All 2 branches missed.">            if (genome[i] &lt; minGene)</span>
<span class="nc" id="L445">                genome[i] = minGene;</span>
            else 
                {
<span class="nc" id="L448">                byte maxGene = (byte)_species.maxGene(i);</span>
<span class="nc bnc" id="L449" title="All 2 branches missed.">                if (genome[i] &gt; maxGene)</span>
<span class="nc" id="L450">                    genome[i] = maxGene;</span>
                }
            }
<span class="nc" id="L453">        }</span>
                
    public void setGenomeLength(int len)
        {
<span class="nc" id="L457">        byte[] newGenome = new byte[len];</span>
<span class="nc" id="L458">        System.arraycopy(genome, 0, newGenome, 0, </span>
<span class="nc bnc" id="L459" title="All 2 branches missed.">            genome.length &lt; newGenome.length ? genome.length : newGenome.length);</span>
<span class="nc" id="L460">        genome = newGenome;</span>
<span class="nc" id="L461">        }</span>

    /** Returns true if each gene value is within is specified [min,max] range. */
    public boolean isInRange() 
        {
<span class="nc" id="L466">        IntegerVectorSpecies _species = (IntegerVectorSpecies)species;</span>
<span class="nc bnc" id="L467" title="All 2 branches missed.">        for (int i = 0; i &lt; genomeLength(); i++)</span>
<span class="nc bnc" id="L468" title="All 2 branches missed.">            if (genome[i] &lt; _species.minGene(i) ||</span>
<span class="nc bnc" id="L469" title="All 2 branches missed.">                genome[i] &gt; _species.maxGene(i)) return false;</span>
<span class="nc" id="L470">        return true;</span>
        }

    public double distanceTo(Individual otherInd)
        {               
<span class="nc bnc" id="L475" title="All 2 branches missed.">        if (!(otherInd instanceof ByteVectorIndividual)) </span>
<span class="nc" id="L476">            return super.distanceTo(otherInd);  // will return infinity!</span>
                
<span class="nc" id="L478">        ByteVectorIndividual other = (ByteVectorIndividual) otherInd;</span>
<span class="nc" id="L479">        byte[] otherGenome = other.genome;</span>
<span class="nc" id="L480">        double sumSquaredDistance =0.0;</span>
<span class="nc bnc" id="L481" title="All 2 branches missed.">        for(int i=0; i &lt; other.genomeLength(); i++)</span>
            {
<span class="nc" id="L483">            long dist = this.genome[i] - (long)otherGenome[i];</span>
<span class="nc" id="L484">            sumSquaredDistance += dist*dist;</span>
            }
<span class="nc" id="L486">        return StrictMath.sqrt(sumSquaredDistance);</span>
        }
    }
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.5.201910111838</span></div></body></html>