<!DOCTYPE HTML>
<html lang="en">
<head>
<!-- Generated by javadoc (18) on Fri Feb 24 15:31:35 EST 2023 -->
<title>CMAESSpecies (ecj 27 API)</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="dc.created" content="2023-02-24">
<meta name="description" content="declaration: package: ec.eda.cmaes, class: CMAESSpecies">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var evenRowColor = "even-row-color";
var oddRowColor = "odd-row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../../../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar-top"><button id="navbar-toggle-button" aria-controls="navbar-top" aria-expanded="false" aria-label="Toggle navigation links"><span class="nav-bar-toggle-icon"></span><span class="nav-bar-toggle-icon"></span><span class="nav-bar-toggle-icon"></span></button>
<div class="skip-nav"><a href="#skip-navbar-top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar-top-firstrow" class="nav-list" title="Navigation">
<li><a href="../../../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="class-use/CMAESSpecies.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-all.html">Index</a></li>
<li><a href="../../../help-doc.html#class">Help</a></li>
</ul>
<ul class="sub-nav-list-small">
<li>
<p>Summary:</p>
<ul>
<li>Nested</li>
<li><a href="#field-summary">Field</a></li>
<li><a href="#constructor-summary">Constr</a></li>
<li><a href="#method-summary">Method</a></li>
</ul>
</li>
<li>
<p>Detail:</p>
<ul>
<li><a href="#field-detail">Field</a></li>
<li><a href="#constructor-detail">Constr</a></li>
<li><a href="#method-detail">Method</a></li>
</ul>
</li>
</ul>
</div>
<div class="sub-nav">
<div id="navbar-sub-list">
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field-summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor-summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method-summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field-detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor-detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method-detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search-input">SEARCH:</label>
<input type="text" id="search-input" disabled placeholder="Search">
<input type="reset" id="reset-button" disabled value="reset">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip-navbar-top"></span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">ec.eda.cmaes</a></div>
<h1 title="Class CMAESSpecies" class="title">Class CMAESSpecies</h1>
</div>
<div class="inheritance" title="Inheritance Tree"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">java.lang.Object</a>
<div class="inheritance"><a href="../../Species.html" title="class in ec">ec.Species</a>
<div class="inheritance"><a href="../../vector/VectorSpecies.html" title="class in ec.vector">ec.vector.VectorSpecies</a>
<div class="inheritance"><a href="../../vector/FloatVectorSpecies.html" title="class in ec.vector">ec.vector.FloatVectorSpecies</a>
<div class="inheritance">ec.eda.cmaes.CMAESSpecies</div>
</div>
</div>
</div>
</div>
<section class="class-description" id="class-description">
<dl class="notes">
<dt>All Implemented Interfaces:</dt>
<dd><code><a href="../../Prototype.html" title="interface in ec">Prototype</a></code>, <code><a href="../../Setup.html" title="interface in ec">Setup</a></code>, <code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/io/Serializable.html" title="class or interface in java.io" class="external-link">Serializable</a></code>, <code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Cloneable.html" title="class or interface in java.lang" class="external-link">Cloneable</a></code></dd>
</dl>
<hr>
<div class="type-signature"><span class="modifiers">public class </span><span class="element-name type-name-label">CMAESSpecies</span>
<span class="extends-implements">extends <a href="../../vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></span></div>
<div class="block">CMAESSpecies is a FloatVectorSpecies which implements a faithful version of the
 CMA-ES algorithm.  The class has two basic methods.  The newIndividual(...)
 method generates a new random individual underneath the current CMA-ES
 gaussian distribution.  The updateDistribution(...) method revises the
 gaussian distribution to reflect the fitness results of the population.
 
 <p>CMAESSpecies must be used in combination with CMAESBreeder, which will
 call it at appropriate times to revise the distribution and to generate a new
 subpopulation of individuals.  It must also be used in combination with
 CMAESInitializer, which will use it to generate the initial population.

 <p>Note importantly that CMAESSpecies <b>ignores the subpopulation size</b>.
 Instead the first thing it will do is revise the subpopulation size to reflect
 the "lambda" parameter.  This is a consequence of another feature of
 CMAESSpecies: many of its parameters do not have fixed default values, but
 rather values which are computed on the fly if the user does not provide them.
 For this reason, it also prints out these values when running so the user may
 see what values it used for that given run.  The computed default values 
 use equations which are common in the CMA-ES literature and are described
 below.

 <p>CMAESSpecies also has an "alternative termination" option.  Normally ECJ
 terminates when the optimal individual is discovered or when the generations
 or maximum number of evaluations has been exceeded.  CMA-ES can optionally
 terminate when the eigenvalues of the covariance matrix of its gaussian
 distribution are too small.  Among other things, this will generally prevent
 CMA-ES from terminating abnormally because an eigenvalue has gone negative
 (due to floating point underflows).  But by default this alternative termination
 is turned off, and CMA-ES will simply terminate in that case.

 <p>CMAESSpecies needs initial values of sigma (the scaling parameter for its
 covariance matrix) and the mean of the gaussian distribution.  By default
 sigma's initial value is 1.0.  The mean must be set to one of "zero"
 (meaning the origin), "center" (meaning the center of the genome space defined
 by the min and max gene values for each gene), or "random" (meaning a randomly-
 chosen point in the space".  If it is not set to any of these, you may
 alternatively set the initial mean values by hand.  But you must do one of
 the two.

 <p>Initializing the covariance matrix can be a problem in in CMA-ES, particularly
 if it is large relative to the gene bounds.  If CMA-ES generates a random individual
 under its current distribution and that individual violates the bounds of just a 
 single gene, it is invalid and must be regenerated.  If you have a lot of genes,
 and the covariance matrix is large relative to their bounds, then the probability
 that this will occur rapidly approaches 1.0, so CMA-ES will be trapped in an effectively
 infinite loop endlessly producing invalid individuals.

 <p>This can be remedied in a few ways.  First there is an option available to force
 the initial covariance matrix to NOT be the identity matrix (the default) but instead
 be scaled according to the gene bounds.  That may help.  You can also of course reduce
 sigma.  Last, you can turn on an alternative individual generation mechanism; here,
 if a specific gene bound is violated, then *for that gene only* the value is chosen at
 random uniformly from within the gene bounds.

 <p>CMAESSpecies relies on the EJML matrix library, available at 
 <a href="http://ejml.org/">http://ejml.org/</a>

 <p><b>Parameters</b><br>
 <table>
 <tr><td valign=top><i>base</i>.<tt>mean</tt><br>
 <font size=-1>String, one of center, zero, or random</font></td>
 <td valign=top>(the initial mean for the distribution)<br>
 If not provided, <i>base</i>.<tt>mean</tt>.0 and so on must be given    
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>mean</tt><br>
 <font size=-1>String, one of center, zero, or random</font></td>
 <td valign=top>(the initial mean for the distribution)<br>
 If not provided, <i>base</i>.<tt>mean</tt>.0 and so on must be given    
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>mean</tt>.<i>i</i><br>
 <font size=-1>Floating-point value</font></td>
 <td valign=top>(the value of dimension i of the initial mean vector)<br>
 These values will override those set as a result of <i>base</i>.<tt>mean</tt> 
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>sigma</tt><br>
 <font size=-1>Floating-point value > 0.0</font></td>
 <td valign=top>(the sigma scaling parameter)<br>
 If not provided, this defaults to 1.0.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>lambda</tt><br>
 <font size=-1>Integer > <i>base</i>.<tt>mu</tt></font></td>
 <td valign=top>(lambda population size)<br>
 If not provided, this defaults to 4 + Math.floor(3 * Math.log(n)).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>lambda</tt><br>
 <font size=-1>Integer > 1</td>
 <td valign=top>(mu truncated population size)<br>
 If not provided, this defaults to Math.floor(lambda / 2.0).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>weight</tt>.<i>i</i><br>
 <font size=-1>Float >= 0</td>
 <td valign=top>(weight for individual i (i from 0 to mu-1))<br>
 If not provided, this defaults to Math.log((lambda + 1.0) / (2.0 * (i + 1))).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>cc</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1</td>
 <td valign=top>(c_c parameter)<br>
 If not provided, this defaults to (4.0+mueff/n) / (n+4.0 + 2.0*mueff/n)<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>
 
 <tr><td valign=top><i>base</i>.<tt>cs</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1</td>
 <td valign=top>(c_sigma parameter)<br>
 If not provided, this defaults to (mueff+2.0)/(n+mueff+5.0)<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>c1</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1 (and c1 > (1-cmu))</td>
 <td valign=top>(c_1 parameter)<br>
 If not provided, this defaults to 2.0 / ((n+1.3)*(n+1.3)+mueff)<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>cmu</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1 (and cmu > (1-c1))</td>
 <td valign=top>(c_mu parameter)<br>
 If not provided, this defaults to Math.min(1.0-c1, 2.0*(mueff-2.0+1.0/mueff) / ((n+2.0)*(n+2.0)+mueff))<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>damps</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1</td>
 <td valign=top>(d_sigma dampening parameter)<br>
 If not provided, this defaults to 1.0 + 2.0*Math.max(0.0, Math.sqrt((mueff-1.0)/(n+1.0))-1.0) + cs<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>covariance</tt><br>
 <font size=-1>String, either "identity" (default) or "scaled"</td>
 <td valign=top>Covariance matrix initialization procedure.<br>
 If "identity", then the covariance matrix is initialized to the
 identity matrix.  If "scaled", then the covariance matrix is
 initialized to a diagonal matrix whose values are squares of
 each gene range (max - min).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>alternative-generator</tt><br>
 <font size=-1>true or false (default)</td>
 <td valign=top>Whether or not to use the alternative indivdiual-generation procedure.<br>
 If "true", then if, in the process of generating an individual, we have failed 
 alternative-generator-tries times to create an individual which falls within the min and
 max gene values for each gene, then whenever a gene value violates those constraints
 we will simply randomize it to something uniformly chosen between the min and max.
 If "false", then whenever an individual violates constraints, we will try again as
 necessary.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>alternative-generator-tries</tt><br>
 <font size=-1>Integer > 1 (default is 100)</td>
 <td valign=top>How many times we try to generate a valid individual before
 possibly using the alternative-generator approach.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>alternative-termination</tt><br>
 <font size=-1>boolean, default false</td>
 <td valign=top>Should we also terminate on CMA-ES's additional termination conditions?.
 </td></tr>

 </table>


 <p><b>Default Base</b><br>
 eda.cma-es.species</div>
<dl class="notes">
<dt>Version:</dt>
<dd>1.0</dd>
<dt>Author:</dt>
<dd>Sam McKay and Sean Luke</dd>
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../serialized-form.html#ec.eda.cmaes.CMAESSpecies">Serialized Form</a></li>
</ul>
</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field-summary">
<h2>Field Summary</h2>
<div class="caption"><span>Fields</span></div>
<div class="summary-table three-column-summary">
<div class="table-header col-first">Modifier and Type</div>
<div class="table-header col-second">Field</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color"><code>int</code></div>
<div class="col-second even-row-color"><code><a href="#altGeneratorTries" class="member-name-link">altGeneratorTries</a></code></div>
<div class="col-last even-row-color">
<div class="block">How many times should we try to generate a valid individual before we give up and use the useAltGenerator approach?</div>
</div>
<div class="col-first odd-row-color"><code>org.ejml.simple.SimpleMatrix</code></div>
<div class="col-second odd-row-color"><code><a href="#b" class="member-name-link">b</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The "B" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</div>
<div class="col-first even-row-color"><code>org.ejml.data.DenseMatrix64F</code></div>
<div class="col-second even-row-color"><code><a href="#bd" class="member-name-link">bd</a></code></div>
<div class="col-last even-row-color">
<div class="block">b x d</div>
</div>
<div class="col-first odd-row-color"><code>org.ejml.simple.SimpleMatrix</code></div>
<div class="col-second odd-row-color"><code><a href="#c" class="member-name-link">c</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The "C" covariance matrix of the distribution.</div>
</div>
<div class="col-first even-row-color"><code>double</code></div>
<div class="col-second even-row-color"><code><a href="#c1" class="member-name-link">c1</a></code></div>
<div class="col-last even-row-color">
<div class="block">The c_1 rank-1 update learning rate.</div>
</div>
<div class="col-first odd-row-color"><code>double</code></div>
<div class="col-second odd-row-color"><code><a href="#cc" class="member-name-link">cc</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The c_c rank-one evolution path cumulation parameter.</div>
</div>
<div class="col-first even-row-color"><code>double</code></div>
<div class="col-second even-row-color"><code><a href="#chiN" class="member-name-link">chiN</a></code></div>
<div class="col-last even-row-color">
<div class="block">An estimate of the expected size of the standard multivariate gaussian N(0,I).</div>
</div>
<div class="col-first odd-row-color"><code>double</code></div>
<div class="col-second odd-row-color"><code><a href="#cmu" class="member-name-link">cmu</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The c_{\mu} rank-mu update learning rate.</div>
</div>
<div class="col-first even-row-color"><code>double</code></div>
<div class="col-second even-row-color"><code><a href="#cs" class="member-name-link">cs</a></code></div>
<div class="col-last even-row-color">
<div class="block">The c_{\sigma} mutation rate evolution path learning rate.</div>
</div>
<div class="col-first odd-row-color"><code>org.ejml.simple.SimpleMatrix</code></div>
<div class="col-second odd-row-color"><code><a href="#d" class="member-name-link">d</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The "C" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</div>
<div class="col-first even-row-color"><code>double</code></div>
<div class="col-second even-row-color"><code><a href="#damps" class="member-name-link">damps</a></code></div>
<div class="col-last even-row-color">
<div class="block">The d_{\sigma} dampening for the mutation rate update.</div>
</div>
<div class="col-first odd-row-color"><code>static final int</code></div>
<div class="col-second odd-row-color"><code><a href="#DEFAULT_ALT_GENERATOR_TRIES" class="member-name-link">DEFAULT_ALT_GENERATOR_TRIES</a></code></div>
<div class="col-last odd-row-color">
<div class="block">Default value (100) for altGeneratorTries.</div>
</div>
<div class="col-first even-row-color"><code>org.ejml.simple.SimpleMatrix</code></div>
<div class="col-second even-row-color"><code><a href="#invsqrtC" class="member-name-link">invsqrtC</a></code></div>
<div class="col-last even-row-color">
<div class="block">C^{-1/2}.</div>
</div>
<div class="col-first odd-row-color"><code>int</code></div>
<div class="col-second odd-row-color"><code><a href="#lambda" class="member-name-link">lambda</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The individuals generated from the distribution.</div>
</div>
<div class="col-first even-row-color"><code>int</code></div>
<div class="col-second even-row-color"><code><a href="#lastEigenDecompositionGeneration" class="member-name-link">lastEigenDecompositionGeneration</a></code></div>
<div class="col-last even-row-color">
<div class="block">The most recent generation where an eigendecomposition on C was performed into B and D</div>
</div>
<div class="col-first odd-row-color"><code>static final int</code></div>
<div class="col-second odd-row-color"><code><a href="#MAX_TRIES_BEFORE_WARNING" class="member-name-link">MAX_TRIES_BEFORE_WARNING</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>int</code></div>
<div class="col-second even-row-color"><code><a href="#mu" class="member-name-link">mu</a></code></div>
<div class="col-last even-row-color">
<div class="block">The truncated individuals used to update the distribution.</div>
</div>
<div class="col-first odd-row-color"><code>double</code></div>
<div class="col-second odd-row-color"><code><a href="#mueff" class="member-name-link">mueff</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The "mu_{eff}" constant in CMA-ES.</div>
</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_ALTERNATIVE_GENERATOR" class="member-name-link">P_ALTERNATIVE_GENERATOR</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_ALTERNATIVE_GENERATOR_TRIES" class="member-name-link">P_ALTERNATIVE_GENERATOR_TRIES</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_ALTERNATIVE_TERMINATION" class="member-name-link">P_ALTERNATIVE_TERMINATION</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_C1" class="member-name-link">P_C1</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_CC" class="member-name-link">P_CC</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_CMAES_SPECIES" class="member-name-link">P_CMAES_SPECIES</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_CMU" class="member-name-link">P_CMU</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_COVARIANCE" class="member-name-link">P_COVARIANCE</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_CS" class="member-name-link">P_CS</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_DAMPS" class="member-name-link">P_DAMPS</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_LAMBDA" class="member-name-link">P_LAMBDA</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_MEAN" class="member-name-link">P_MEAN</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_MU" class="member-name-link">P_MU</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#P_SIGMA" class="member-name-link">P_SIGMA</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#P_WEIGHTS" class="member-name-link">P_WEIGHTS</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>org.ejml.simple.SimpleMatrix</code></div>
<div class="col-second odd-row-color"><code><a href="#pc" class="member-name-link">pc</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The p_c evolution path vector.</div>
</div>
<div class="col-first even-row-color"><code>org.ejml.simple.SimpleMatrix</code></div>
<div class="col-second even-row-color"><code><a href="#ps" class="member-name-link">ps</a></code></div>
<div class="col-last even-row-color">
<div class="block">The p_{\sigma} evolution path vector.</div>
</div>
<div class="col-first odd-row-color"><code>org.ejml.data.DenseMatrix64F</code></div>
<div class="col-second odd-row-color"><code><a href="#sbd" class="member-name-link">sbd</a></code></div>
<div class="col-last odd-row-color">
<div class="block">bd x sigma</div>
</div>
<div class="col-first even-row-color"><code>double</code></div>
<div class="col-second even-row-color"><code><a href="#sigma" class="member-name-link">sigma</a></code></div>
<div class="col-last even-row-color">
<div class="block">The "sigma" scaling factor for the covariance matrix.</div>
</div>
<div class="col-first odd-row-color"><code>boolean</code></div>
<div class="col-second odd-row-color"><code><a href="#useAltGenerator" class="member-name-link">useAltGenerator</a></code></div>
<div class="col-last odd-row-color">
<div class="block">If, after trying altGeneratorTries to build an indiviual, we are still building one which violates min/max gene constraints, should
        we instead fill those violated genes with uniformly-selected values between the min and max?</div>
</div>
<div class="col-first even-row-color"><code>boolean</code></div>
<div class="col-second even-row-color"><code><a href="#useAltTermination" class="member-name-link">useAltTermination</a></code></div>
<div class="col-last even-row-color">
<div class="block">Should we terminate when the eigenvalues get too small?  If we don't, they might go negative and the eigendecomposition will fail.</div>
</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#V_CENTER" class="member-name-link">V_CENTER</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#V_IDENTITY" class="member-name-link">V_IDENTITY</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#V_RANDOM" class="member-name-link">V_RANDOM</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second even-row-color"><code><a href="#V_SCALED" class="member-name-link">V_SCALED</a></code></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-first odd-row-color"><code>static final <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></code></div>
<div class="col-second odd-row-color"><code><a href="#V_ZERO" class="member-name-link">V_ZERO</a></code></div>
<div class="col-last odd-row-color">&nbsp;</div>
<div class="col-first even-row-color"><code>double[]</code></div>
<div class="col-second even-row-color"><code><a href="#weights" class="member-name-link">weights</a></code></div>
<div class="col-last even-row-color">
<div class="block">The ranked fitness weights for the mu individuals.</div>
</div>
<div class="col-first odd-row-color"><code>org.ejml.simple.SimpleMatrix</code></div>
<div class="col-second odd-row-color"><code><a href="#xmean" class="member-name-link">xmean</a></code></div>
<div class="col-last odd-row-color">
<div class="block">The mean of the distribution.</div>
</div>
</div>
<div class="inherited-list">
<h3 id="fields-inherited-from-class-ec.vector.FloatVectorSpecies">Fields inherited from class&nbsp;ec.vector.<a href="../../vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></h3>
<code><a href="../../vector/FloatVectorSpecies.html#C_GAUSS_MUTATION">C_GAUSS_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#C_INTEGER_RANDOM_WALK_MUTATION">C_INTEGER_RANDOM_WALK_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#C_INTEGER_RESET_MUTATION">C_INTEGER_RESET_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#C_POLYNOMIAL_MUTATION">C_POLYNOMIAL_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#C_RESET_MUTATION">C_RESET_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#DEFAULT_OUT_OF_BOUNDS_RETRIES">DEFAULT_OUT_OF_BOUNDS_RETRIES</a>, <a href="../../vector/FloatVectorSpecies.html#gaussMutationStdev">gaussMutationStdev</a>, <a href="../../vector/FloatVectorSpecies.html#maxGene">maxGene</a>, <a href="../../vector/FloatVectorSpecies.html#minGene">minGene</a>, <a href="../../vector/FloatVectorSpecies.html#mutationDistributionIndex">mutationDistributionIndex</a>, <a href="../../vector/FloatVectorSpecies.html#mutationIsBounded">mutationIsBounded</a>, <a href="../../vector/FloatVectorSpecies.html#mutationType">mutationType</a>, <a href="../../vector/FloatVectorSpecies.html#outOfBoundsRetries">outOfBoundsRetries</a>, <a href="../../vector/FloatVectorSpecies.html#P_MAXGENE">P_MAXGENE</a>, <a href="../../vector/FloatVectorSpecies.html#P_MINGENE">P_MINGENE</a>, <a href="../../vector/FloatVectorSpecies.html#P_MUTATION_BOUNDED">P_MUTATION_BOUNDED</a>, <a href="../../vector/FloatVectorSpecies.html#P_MUTATION_DISTRIBUTION_INDEX">P_MUTATION_DISTRIBUTION_INDEX</a>, <a href="../../vector/FloatVectorSpecies.html#P_MUTATIONTYPE">P_MUTATIONTYPE</a>, <a href="../../vector/FloatVectorSpecies.html#P_OUTOFBOUNDS_RETRIES">P_OUTOFBOUNDS_RETRIES</a>, <a href="../../vector/FloatVectorSpecies.html#P_POLYNOMIAL_ALTERNATIVE">P_POLYNOMIAL_ALTERNATIVE</a>, <a href="../../vector/FloatVectorSpecies.html#P_RANDOM_WALK_PROBABILITY">P_RANDOM_WALK_PROBABILITY</a>, <a href="../../vector/FloatVectorSpecies.html#P_STDEV">P_STDEV</a>, <a href="../../vector/FloatVectorSpecies.html#polynomialIsAlternative">polynomialIsAlternative</a>, <a href="../../vector/FloatVectorSpecies.html#randomWalkProbability">randomWalkProbability</a>, <a href="../../vector/FloatVectorSpecies.html#V_GAUSS_MUTATION">V_GAUSS_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#V_INTEGER_RANDOM_WALK_MUTATION">V_INTEGER_RANDOM_WALK_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#V_INTEGER_RESET_MUTATION">V_INTEGER_RESET_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#V_POLYNOMIAL_MUTATION">V_POLYNOMIAL_MUTATION</a>, <a href="../../vector/FloatVectorSpecies.html#V_RESET_MUTATION">V_RESET_MUTATION</a></code></div>
<div class="inherited-list">
<h3 id="fields-inherited-from-class-ec.vector.VectorSpecies">Fields inherited from class&nbsp;ec.vector.<a href="../../vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></h3>
<code><a href="../../vector/VectorSpecies.html#C_ANY_POINT">C_ANY_POINT</a>, <a href="../../vector/VectorSpecies.html#C_GEOMETRIC">C_GEOMETRIC</a>, <a href="../../vector/VectorSpecies.html#C_INTERMED_RECOMB">C_INTERMED_RECOMB</a>, <a href="../../vector/VectorSpecies.html#C_LINE_RECOMB">C_LINE_RECOMB</a>, <a href="../../vector/VectorSpecies.html#C_NONE">C_NONE</a>, <a href="../../vector/VectorSpecies.html#C_ONE_POINT">C_ONE_POINT</a>, <a href="../../vector/VectorSpecies.html#C_ONE_POINT_NO_NOP">C_ONE_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#C_SIMULATED_BINARY">C_SIMULATED_BINARY</a>, <a href="../../vector/VectorSpecies.html#C_TWO_POINT">C_TWO_POINT</a>, <a href="../../vector/VectorSpecies.html#C_TWO_POINT_NO_NOP">C_TWO_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#C_UNIFORM">C_UNIFORM</a>, <a href="../../vector/VectorSpecies.html#chunksize">chunksize</a>, <a href="../../vector/VectorSpecies.html#crossoverDistributionIndex">crossoverDistributionIndex</a>, <a href="../../vector/VectorSpecies.html#crossoverProbability">crossoverProbability</a>, <a href="../../vector/VectorSpecies.html#crossoverType">crossoverType</a>, <a href="../../vector/VectorSpecies.html#duplicateRetries">duplicateRetries</a>, <a href="../../vector/VectorSpecies.html#dynamicInitialSize">dynamicInitialSize</a>, <a href="../../vector/VectorSpecies.html#genomeIncreaseProbability">genomeIncreaseProbability</a>, <a href="../../vector/VectorSpecies.html#genomeResizeAlgorithm">genomeResizeAlgorithm</a>, <a href="../../vector/VectorSpecies.html#genomeSize">genomeSize</a>, <a href="../../vector/VectorSpecies.html#lineDistance">lineDistance</a>, <a href="../../vector/VectorSpecies.html#maxInitialSize">maxInitialSize</a>, <a href="../../vector/VectorSpecies.html#minInitialSize">minInitialSize</a>, <a href="../../vector/VectorSpecies.html#mutationProbability">mutationProbability</a>, <a href="../../vector/VectorSpecies.html#P_CHUNKSIZE">P_CHUNKSIZE</a>, <a href="../../vector/VectorSpecies.html#P_CROSSOVER_DISTRIBUTION_INDEX">P_CROSSOVER_DISTRIBUTION_INDEX</a>, <a href="../../vector/VectorSpecies.html#P_CROSSOVERPROB">P_CROSSOVERPROB</a>, <a href="../../vector/VectorSpecies.html#P_CROSSOVERTYPE">P_CROSSOVERTYPE</a>, <a href="../../vector/VectorSpecies.html#P_DUPLICATE_RETRIES">P_DUPLICATE_RETRIES</a>, <a href="../../vector/VectorSpecies.html#P_GENOMESIZE">P_GENOMESIZE</a>, <a href="../../vector/VectorSpecies.html#P_GEOMETRIC_PROBABILITY">P_GEOMETRIC_PROBABILITY</a>, <a href="../../vector/VectorSpecies.html#P_LINEDISTANCE">P_LINEDISTANCE</a>, <a href="../../vector/VectorSpecies.html#P_MUTATIONPROB">P_MUTATIONPROB</a>, <a href="../../vector/VectorSpecies.html#P_NUM_SEGMENTS">P_NUM_SEGMENTS</a>, <a href="../../vector/VectorSpecies.html#P_SEGMENT">P_SEGMENT</a>, <a href="../../vector/VectorSpecies.html#P_SEGMENT_END">P_SEGMENT_END</a>, <a href="../../vector/VectorSpecies.html#P_SEGMENT_START">P_SEGMENT_START</a>, <a href="../../vector/VectorSpecies.html#P_SEGMENT_TYPE">P_SEGMENT_TYPE</a>, <a href="../../vector/VectorSpecies.html#P_UNIFORM_MAX">P_UNIFORM_MAX</a>, <a href="../../vector/VectorSpecies.html#P_UNIFORM_MIN">P_UNIFORM_MIN</a>, <a href="../../vector/VectorSpecies.html#P_VECTORSPECIES">P_VECTORSPECIES</a>, <a href="../../vector/VectorSpecies.html#V_ANY_POINT">V_ANY_POINT</a>, <a href="../../vector/VectorSpecies.html#V_GEOMETRIC">V_GEOMETRIC</a>, <a href="../../vector/VectorSpecies.html#V_INTERMED_RECOMB">V_INTERMED_RECOMB</a>, <a href="../../vector/VectorSpecies.html#V_LINE_RECOMB">V_LINE_RECOMB</a>, <a href="../../vector/VectorSpecies.html#V_ONE_POINT">V_ONE_POINT</a>, <a href="../../vector/VectorSpecies.html#V_ONE_POINT_NO_NOP">V_ONE_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#V_SIMULATED_BINARY">V_SIMULATED_BINARY</a>, <a href="../../vector/VectorSpecies.html#V_TWO_POINT">V_TWO_POINT</a>, <a href="../../vector/VectorSpecies.html#V_TWO_POINT_NO_NOP">V_TWO_POINT_NO_NOP</a>, <a href="../../vector/VectorSpecies.html#V_UNIFORM">V_UNIFORM</a></code></div>
<div class="inherited-list">
<h3 id="fields-inherited-from-class-ec.Species">Fields inherited from class&nbsp;ec.<a href="../../Species.html" title="class in ec">Species</a></h3>
<code><a href="../../Species.html#f_prototype">f_prototype</a>, <a href="../../Species.html#i_prototype">i_prototype</a>, <a href="../../Species.html#P_FITNESS">P_FITNESS</a>, <a href="../../Species.html#P_INDIVIDUAL">P_INDIVIDUAL</a>, <a href="../../Species.html#P_PIPE">P_PIPE</a>, <a href="../../Species.html#pipe_prototype">pipe_prototype</a></code></div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor-summary">
<h2>Constructor Summary</h2>
<div class="caption"><span>Constructors</span></div>
<div class="summary-table two-column-summary">
<div class="table-header col-first">Constructor</div>
<div class="table-header col-last">Description</div>
<div class="col-constructor-name even-row-color"><code><a href="#%3Cinit%3E()" class="member-name-link">CMAESSpecies</a>()</code></div>
<div class="col-last even-row-color">&nbsp;</div>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method-summary">
<h2>Method Summary</h2>
<div id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button id="method-summary-table-tab0" role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table', 3)" class="active-table-tab">All Methods</button><button id="method-summary-table-tab2" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab2', 3)" class="table-tab">Instance Methods</button><button id="method-summary-table-tab4" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab4', 3)" class="table-tab">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<div class="summary-table three-column-summary" aria-labelledby="method-summary-table-tab0">
<div class="table-header col-first">Modifier and Type</div>
<div class="table-header col-second">Method</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#clone()" class="member-name-link">clone</a>()</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="../../util/Parameter.html" title="class in ec.util">Parameter</a></code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#defaultBase()" class="member-name-link">defaultBase</a>()</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">Returns the default base for this prototype.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="../../Individual.html" title="class in ec">Individual</a></code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#newIndividual(ec.EvolutionState,int)" class="member-name-link">newIndividual</a><wbr>(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 int&nbsp;thread)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">Provides a brand-new individual to fill in a population.</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code>void</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#setup(ec.EvolutionState,ec.util.Parameter)" class="member-name-link">setup</a><wbr>(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code>void</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4"><code><a href="#updateDistribution(ec.EvolutionState,ec.Subpopulation)" class="member-name-link">updateDistribution</a><wbr>(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../Subpopulation.html" title="class in ec">Subpopulation</a>&nbsp;subpop)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab4">
<div class="block">Revises the CMA-ES distribution to reflect the current fitness results in the provided subpopulation.</div>
</div>
</div>
</div>
</div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-ec.vector.FloatVectorSpecies">Methods inherited from class&nbsp;ec.vector.<a href="../../vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></h3>
<code><a href="../../vector/FloatVectorSpecies.html#gaussMutationStdev(int)">gaussMutationStdev</a>, <a href="../../vector/FloatVectorSpecies.html#inNumericalTypeRange(double)">inNumericalTypeRange</a>, <a href="../../vector/FloatVectorSpecies.html#loadParametersForGene(ec.EvolutionState,int,ec.util.Parameter,ec.util.Parameter,java.lang.String)">loadParametersForGene</a>, <a href="../../vector/FloatVectorSpecies.html#maxGene(int)">maxGene</a>, <a href="../../vector/FloatVectorSpecies.html#minGene(int)">minGene</a>, <a href="../../vector/FloatVectorSpecies.html#mutationDistributionIndex(int)">mutationDistributionIndex</a>, <a href="../../vector/FloatVectorSpecies.html#mutationIsBounded(int)">mutationIsBounded</a>, <a href="../../vector/FloatVectorSpecies.html#mutationType(int)">mutationType</a>, <a href="../../vector/FloatVectorSpecies.html#outOfRangeRetryLimitReached(ec.EvolutionState)">outOfRangeRetryLimitReached</a>, <a href="../../vector/FloatVectorSpecies.html#polynomialIsAlternative(int)">polynomialIsAlternative</a>, <a href="../../vector/FloatVectorSpecies.html#randomWalkProbability(int)">randomWalkProbability</a></code></div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-ec.vector.VectorSpecies">Methods inherited from class&nbsp;ec.vector.<a href="../../vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></h3>
<code><a href="../../vector/VectorSpecies.html#contains(boolean%5B%5D,boolean)">contains</a>, <a href="../../vector/VectorSpecies.html#contains(double%5B%5D,double)">contains</a>, <a href="../../vector/VectorSpecies.html#contains(int%5B%5D,int)">contains</a>, <a href="../../vector/VectorSpecies.html#contains(long%5B%5D,long)">contains</a>, <a href="../../vector/VectorSpecies.html#duplicateRetries(int)">duplicateRetries</a>, <a href="../../vector/VectorSpecies.html#fill(boolean%5B%5D,boolean)">fill</a>, <a href="../../vector/VectorSpecies.html#fill(double%5B%5D,double)">fill</a>, <a href="../../vector/VectorSpecies.html#fill(int%5B%5D,int)">fill</a>, <a href="../../vector/VectorSpecies.html#fill(long%5B%5D,long)">fill</a>, <a href="../../vector/VectorSpecies.html#initializeGenomeSegmentsByEndIndices(ec.EvolutionState,ec.util.Parameter,ec.util.Parameter,int)">initializeGenomeSegmentsByEndIndices</a>, <a href="../../vector/VectorSpecies.html#initializeGenomeSegmentsByStartIndices(ec.EvolutionState,ec.util.Parameter,ec.util.Parameter,int)">initializeGenomeSegmentsByStartIndices</a>, <a href="../../vector/VectorSpecies.html#mutationProbability(int)">mutationProbability</a>, <a href="../../vector/VectorSpecies.html#setupGenome(ec.EvolutionState,ec.util.Parameter)">setupGenome</a></code></div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-ec.Species">Methods inherited from class&nbsp;ec.<a href="../../Species.html" title="class in ec">Species</a></h3>
<code><a href="../../Species.html#buildMisc(ec.EvolutionState,int,int)">buildMisc</a>, <a href="../../Species.html#newIndividual(ec.EvolutionState,java.io.DataInput)">newIndividual</a>, <a href="../../Species.html#newIndividual(ec.EvolutionState,java.io.LineNumberReader)">newIndividual</a>, <a href="../../Species.html#updateIndividual(ec.EvolutionState,ec.Individual)">updateIndividual</a>, <a href="../../Species.html#updateSubpopulation(ec.EvolutionState,ec.Subpopulation)">updateSubpopulation</a></code></div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-java.lang.Object">Methods inherited from class&nbsp;java.lang.<a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></h3>
<code><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#equals(java.lang.Object)" title="class or interface in java.lang" class="external-link">equals</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#finalize()" title="class or interface in java.lang" class="external-link">finalize</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#getClass()" title="class or interface in java.lang" class="external-link">getClass</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#hashCode()" title="class or interface in java.lang" class="external-link">hashCode</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#notify()" title="class or interface in java.lang" class="external-link">notify</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#notifyAll()" title="class or interface in java.lang" class="external-link">notifyAll</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#toString()" title="class or interface in java.lang" class="external-link">toString</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#wait()" title="class or interface in java.lang" class="external-link">wait</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#wait(long)" title="class or interface in java.lang" class="external-link">wait</a>, <a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html#wait(long,int)" title="class or interface in java.lang" class="external-link">wait</a></code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field-detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="P_CMAES_SPECIES">
<h3>P_CMAES_SPECIES</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_CMAES_SPECIES</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CMAES_SPECIES">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_LAMBDA">
<h3>P_LAMBDA</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_LAMBDA</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_LAMBDA">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_MU">
<h3>P_MU</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_MU</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_MU">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_SIGMA">
<h3>P_SIGMA</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_SIGMA</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_SIGMA">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_MEAN">
<h3>P_MEAN</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_MEAN</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_MEAN">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_WEIGHTS">
<h3>P_WEIGHTS</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_WEIGHTS</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_WEIGHTS">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_CC">
<h3>P_CC</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_CC</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CC">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_CS">
<h3>P_CS</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_CS</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CS">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_C1">
<h3>P_C1</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_C1</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_C1">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_CMU">
<h3>P_CMU</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_CMU</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CMU">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_DAMPS">
<h3>P_DAMPS</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_DAMPS</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_DAMPS">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="V_CENTER">
<h3>V_CENTER</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">V_CENTER</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_CENTER">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="V_ZERO">
<h3>V_ZERO</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">V_ZERO</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_ZERO">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="V_RANDOM">
<h3>V_RANDOM</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">V_RANDOM</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_RANDOM">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_COVARIANCE">
<h3>P_COVARIANCE</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_COVARIANCE</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_COVARIANCE">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="V_IDENTITY">
<h3>V_IDENTITY</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">V_IDENTITY</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_IDENTITY">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="V_SCALED">
<h3>V_SCALED</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">V_SCALED</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_SCALED">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_ALTERNATIVE_TERMINATION">
<h3>P_ALTERNATIVE_TERMINATION</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_ALTERNATIVE_TERMINATION</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_ALTERNATIVE_TERMINATION">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_ALTERNATIVE_GENERATOR">
<h3>P_ALTERNATIVE_GENERATOR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_ALTERNATIVE_GENERATOR</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_ALTERNATIVE_GENERATOR">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="P_ALTERNATIVE_GENERATOR_TRIES">
<h3>P_ALTERNATIVE_GENERATOR_TRIES</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a></span>&nbsp;<span class="element-name">P_ALTERNATIVE_GENERATOR_TRIES</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_ALTERNATIVE_GENERATOR_TRIES">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="lambda">
<h3>lambda</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">lambda</span></div>
<div class="block">The individuals generated from the distribution. 
        If not specified in the parameters, by default 
        lambda = 4+(int)Math.floor(3*Math.log(n));</div>
</section>
</li>
<li>
<section class="detail" id="mu">
<h3>mu</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">mu</span></div>
<div class="block">The truncated individuals used to update the distribution. 
        If not specified in the parameters, by default 
        mu = (int)Math.floor(lambda/2.0);</div>
</section>
</li>
<li>
<section class="detail" id="weights">
<h3>weights</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double[]</span>&nbsp;<span class="element-name">weights</span></div>
<div class="block">The ranked fitness weights for the mu individuals. 
        If not specified in the parameters, by default 
        weights[i] = ln((lambda + 1) / 2i).
        Then all the weights are normalized so that they sum to 1.</div>
</section>
</li>
<li>
<section class="detail" id="mueff">
<h3>mueff</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">mueff</span></div>
<div class="block">The "mu_{eff}" constant in CMA-ES.   This is set to 
        1.0 / the sum of the squares of each of the weights[...]</div>
</section>
</li>
<li>
<section class="detail" id="cmu">
<h3>cmu</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">cmu</span></div>
<div class="block">The c_{\mu} rank-mu update learning rate.
        If not specified in the parameters, by default
        cmu = Math.min(1-c1, 2 * (mueff - 2 + 1/mueff) / ((n+2)*(n+2) + mueff))
        where n is the genome size.</div>
</section>
</li>
<li>
<section class="detail" id="c1">
<h3>c1</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">c1</span></div>
<div class="block">The c_1 rank-1 update learning rate.
        If not specified in the parameters, by default
        c1 = 2 / ((n + 1.3) * (n + 1.3) + mueff)
        where n is the genome size.</div>
</section>
</li>
<li>
<section class="detail" id="cc">
<h3>cc</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">cc</span></div>
<div class="block">The c_c rank-one evolution path cumulation parameter.
        If not specified in the parameters, by default
        cc = (4 + mueff / n) / (n + 4 + 2 * mueff/n)
        where n is the genome size.</div>
</section>
</li>
<li>
<section class="detail" id="cs">
<h3>cs</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">cs</span></div>
<div class="block">The c_{\sigma} mutation rate evolution path learning rate.
        If not specified in the parameters, by default
        cs = (mueff + 2) / (n + mueff + 5)
        where n is the genome size.</div>
</section>
</li>
<li>
<section class="detail" id="damps">
<h3>damps</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">damps</span></div>
<div class="block">The d_{\sigma} dampening for the mutation rate update.
        If not specified in the parameters, by default
        damps = cs + 2 * Math.max(1, Math.sqrt((mueff - 1) / (n + 1))) - 1
        otherwise known as
        damps = 1 + 2 * Math.max(0, Math.sqrt((mueff - 1) / (n + 1)) - 1) + cs
        where n is the genome size.</div>
</section>
</li>
<li>
<section class="detail" id="sigma">
<h3>sigma</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">sigma</span></div>
<div class="block">The "sigma" scaling factor for the covariance matrix.</div>
</section>
</li>
<li>
<section class="detail" id="xmean">
<h3>xmean</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.simple.SimpleMatrix</span>&nbsp;<span class="element-name">xmean</span></div>
<div class="block">The mean of the distribution.</div>
</section>
</li>
<li>
<section class="detail" id="c">
<h3>c</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.simple.SimpleMatrix</span>&nbsp;<span class="element-name">c</span></div>
<div class="block">The "C" covariance matrix of the distribution.</div>
</section>
</li>
<li>
<section class="detail" id="b">
<h3>b</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.simple.SimpleMatrix</span>&nbsp;<span class="element-name">b</span></div>
<div class="block">The "B" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</section>
</li>
<li>
<section class="detail" id="d">
<h3>d</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.simple.SimpleMatrix</span>&nbsp;<span class="element-name">d</span></div>
<div class="block">The "C" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</section>
</li>
<li>
<section class="detail" id="bd">
<h3>bd</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.data.DenseMatrix64F</span>&nbsp;<span class="element-name">bd</span></div>
<div class="block">b x d</div>
</section>
</li>
<li>
<section class="detail" id="sbd">
<h3>sbd</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.data.DenseMatrix64F</span>&nbsp;<span class="element-name">sbd</span></div>
<div class="block">bd x sigma</div>
</section>
</li>
<li>
<section class="detail" id="invsqrtC">
<h3>invsqrtC</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.simple.SimpleMatrix</span>&nbsp;<span class="element-name">invsqrtC</span></div>
<div class="block">C^{-1/2}.  This is equal to B x D^{-1} x B^T</div>
</section>
</li>
<li>
<section class="detail" id="ps">
<h3>ps</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.simple.SimpleMatrix</span>&nbsp;<span class="element-name">ps</span></div>
<div class="block">The p_{\sigma} evolution path vector.</div>
</section>
</li>
<li>
<section class="detail" id="pc">
<h3>pc</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">org.ejml.simple.SimpleMatrix</span>&nbsp;<span class="element-name">pc</span></div>
<div class="block">The p_c evolution path vector.</div>
</section>
</li>
<li>
<section class="detail" id="chiN">
<h3>chiN</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">chiN</span></div>
<div class="block">An estimate of the expected size of the standard multivariate gaussian N(0,I). 
        This is chiN = Math.sqrt(n)*(1.0-1.0/(4.0*n)+1.0/(21.0*n*n))</div>
</section>
</li>
<li>
<section class="detail" id="lastEigenDecompositionGeneration">
<h3>lastEigenDecompositionGeneration</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">lastEigenDecompositionGeneration</span></div>
<div class="block">The most recent generation where an eigendecomposition on C was performed into B and D</div>
</section>
</li>
<li>
<section class="detail" id="useAltTermination">
<h3>useAltTermination</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">boolean</span>&nbsp;<span class="element-name">useAltTermination</span></div>
<div class="block">Should we terminate when the eigenvalues get too small?  If we don't, they might go negative and the eigendecomposition will fail.</div>
</section>
</li>
<li>
<section class="detail" id="useAltGenerator">
<h3>useAltGenerator</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">boolean</span>&nbsp;<span class="element-name">useAltGenerator</span></div>
<div class="block">If, after trying altGeneratorTries to build an indiviual, we are still building one which violates min/max gene constraints, should
        we instead fill those violated genes with uniformly-selected values between the min and max?</div>
</section>
</li>
<li>
<section class="detail" id="altGeneratorTries">
<h3>altGeneratorTries</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">altGeneratorTries</span></div>
<div class="block">How many times should we try to generate a valid individual before we give up and use the useAltGenerator approach?</div>
</section>
</li>
<li>
<section class="detail" id="DEFAULT_ALT_GENERATOR_TRIES">
<h3>DEFAULT_ALT_GENERATOR_TRIES</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">DEFAULT_ALT_GENERATOR_TRIES</span></div>
<div class="block">Default value (100) for altGeneratorTries.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.DEFAULT_ALT_GENERATOR_TRIES">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="MAX_TRIES_BEFORE_WARNING">
<h3>MAX_TRIES_BEFORE_WARNING</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">MAX_TRIES_BEFORE_WARNING</span></div>
<dl class="notes">
<dt>See Also:</dt>
<dd>
<ul class="see-list">
<li><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.MAX_TRIES_BEFORE_WARNING">Constant Field Values</a></li>
</ul>
</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor-detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;()">
<h3>CMAESSpecies</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="element-name">CMAESSpecies</span>()</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method-detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="defaultBase()">
<h3>defaultBase</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type"><a href="../../util/Parameter.html" title="class in ec.util">Parameter</a></span>&nbsp;<span class="element-name">defaultBase</span>()</div>
<div class="block"><span class="description-from-type-label">Description copied from interface:&nbsp;<code><a href="../../Prototype.html#defaultBase()">Prototype</a></code></span></div>
<div class="block">Returns the default base for this prototype.
        This should generally be implemented by building off of the static base()
        method on the DefaultsForm object for the prototype's package. This should
        be callable during setup(...).</div>
<dl class="notes">
<dt>Specified by:</dt>
<dd><code><a href="../../Prototype.html#defaultBase()">defaultBase</a></code>&nbsp;in interface&nbsp;<code><a href="../../Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt>Overrides:</dt>
<dd><code><a href="../../vector/VectorSpecies.html#defaultBase()">defaultBase</a></code>&nbsp;in class&nbsp;<code><a href="../../vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></code></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="setup(ec.EvolutionState,ec.util.Parameter)">
<h3>setup</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name">setup</span><wbr><span class="parameters">(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</span></div>
<div class="block"><span class="description-from-type-label">Description copied from class:&nbsp;<code><a href="../../Species.html#setup(ec.EvolutionState,ec.util.Parameter)">Species</a></code></span></div>
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.  
        If your individual prototype might need to know special things about the species (like parameters stored in it),
        then when you override this setup method, you'll need to set those parameters BEFORE you call super.setup(...),
        because the setup(...) code in Species sets up the prototype.</div>
<dl class="notes">
<dt>Specified by:</dt>
<dd><code><a href="../../Prototype.html#setup(ec.EvolutionState,ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt>Specified by:</dt>
<dd><code><a href="../../Setup.html#setup(ec.EvolutionState,ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../Setup.html" title="interface in ec">Setup</a></code></dd>
<dt>Overrides:</dt>
<dd><code><a href="../../vector/FloatVectorSpecies.html#setup(ec.EvolutionState,ec.util.Parameter)">setup</a></code>&nbsp;in class&nbsp;<code><a href="../../vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></code></dd>
<dt>See Also:</dt>
<dd>
<ul class="see-list-long">
<li><a href="../../Prototype.html#setup(ec.EvolutionState,ec.util.Parameter)"><code>Prototype.setup(EvolutionState,Parameter)</code></a></li>
</ul>
</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="clone()">
<h3>clone</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type"><a href="https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></span>&nbsp;<span class="element-name">clone</span>()</div>
<div class="block"><span class="description-from-type-label">Description copied from interface:&nbsp;<code><a href="../../Prototype.html#clone()">Prototype</a></code></span></div>
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.

        <p>Typically this should be a full "deep" clone.
        However, you may share certain elements with other objects
        rather than clone hem, depending on the situation:

        <p>
        <ul>
        <li>If you hold objects which are shared with other instances,
        don't clone them.
        <li>If you hold objects which must be unique, clone them.
        <li>If you hold objects which were given to you as a gesture
        of kindness, and aren't owned by you, you probably shouldn't clone
        them.
        <li> DON'T attempt to clone: Singletons, Cliques, or Populations, or Subpopulation.
        <li>Arrays are not cloned automatically; you may need to
        clone an array if you're not sharing it with other instances.
        Arrays have the nice feature of being copyable by calling clone()
        on them.
        </ul>

        <p><b>Implementations.</b>

        <ul>
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are abstract, then you should not declare clone().
        
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are <b>not</b> abstract, then you should implement
        it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         return super.clone();
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     }
        </pre></tt>
        
        <li>If no ancestor of yours implements clone(), but you
        need to deep-clone some things, then you should implement it
        as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         MyObject myobj = (MyObject) (super.clone());

         // put your deep-cloning code here...
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     return myobj;
     } 
        </pre></tt>

        <li>If an ancestor has implemented clone(), and you also need
        to deep clone some things, then you should implement it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     { 
     MyObject myobj = (MyObject) (super.clone());

     // put your deep-cloning code here...

     return myobj;
     } 
        </pre></tt>
        </ul></div>
<dl class="notes">
<dt>Specified by:</dt>
<dd><code><a href="../../Prototype.html#clone()">clone</a></code>&nbsp;in interface&nbsp;<code><a href="../../Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt>Overrides:</dt>
<dd><code><a href="../../Species.html#clone()">clone</a></code>&nbsp;in class&nbsp;<code><a href="../../Species.html" title="class in ec">Species</a></code></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="newIndividual(ec.EvolutionState,int)">
<h3>newIndividual</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type"><a href="../../Individual.html" title="class in ec">Individual</a></span>&nbsp;<span class="element-name">newIndividual</span><wbr><span class="parameters">(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 int&nbsp;thread)</span></div>
<div class="block"><span class="description-from-type-label">Description copied from class:&nbsp;<code><a href="../../Species.html#newIndividual(ec.EvolutionState,int)">Species</a></code></span></div>
<div class="block">Provides a brand-new individual to fill in a population.  The default form
        simply calls clone(), creates a fitness, sets evaluated to false, and sets
        the species.  If you need to make a more custom genotype (as is the case
        for GPSpecies, which requires a light rather than deep clone), 
        you will need to override this method as you see fit.</div>
<dl class="notes">
<dt>Overrides:</dt>
<dd><code><a href="../../vector/VectorSpecies.html#newIndividual(ec.EvolutionState,int)">newIndividual</a></code>&nbsp;in class&nbsp;<code><a href="../../vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></code></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="updateDistribution(ec.EvolutionState,ec.Subpopulation)">
<h3>updateDistribution</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name">updateDistribution</span><wbr><span class="parameters">(<a href="../../EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
 <a href="../../Subpopulation.html" title="class in ec">Subpopulation</a>&nbsp;subpop)</span></div>
<div class="block">Revises the CMA-ES distribution to reflect the current fitness results in the provided subpopulation.</div>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<hr>
<p class="legal-copy"><small>Copyright &#169; 2023. All rights reserved.</small></p>
</footer>
</div>
</div>
</body>
</html>
